{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ba544983",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Main model loaded\n",
      "Input 'aa' tensor shape: torch.Size([1, 110])\n",
      "Input 'mask' tensor shape: torch.Size([1, 110])\n"
     ]
    }
   ],
   "source": [
    "import typing as T\n",
    "from dataclasses import dataclass\n",
    "from functools import partial\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from torch import nn\n",
    "from torch.nn import LayerNorm\n",
    "\n",
    "from data import Alphabet\n",
    "from categorical_mixture import categorical_lddt\n",
    "from trunk import FoldingTrunk, FoldingTrunkConfig\n",
    "import residue_constants\n",
    "\n",
    "from typing import Dict, Optional, Tuple\n",
    "\n",
    "SEQ_FILE = \"protein_seq_short.txt\"  # File you just created\n",
    "DEVICE = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "def _calculate_bin_centers(boundaries: torch.Tensor):\n",
    "    step = boundaries[1] - boundaries[0]\n",
    "    bin_centers = boundaries + step / 2\n",
    "    bin_centers = torch.cat(\n",
    "        [bin_centers, (bin_centers[-1] + step).unsqueeze(-1)], dim=0\n",
    "    )\n",
    "    return bin_centers\n",
    "\n",
    "def _calculate_expected_aligned_error(\n",
    "    alignment_confidence_breaks: torch.Tensor,\n",
    "    aligned_distance_error_probs: torch.Tensor,\n",
    ") -> Tuple[torch.Tensor, torch.Tensor]:\n",
    "    bin_centers = _calculate_bin_centers(alignment_confidence_breaks)\n",
    "    return (\n",
    "        torch.sum(aligned_distance_error_probs * bin_centers, dim=-1),\n",
    "        bin_centers[-1],\n",
    "    )\n",
    "\n",
    "def compute_tm(\n",
    "    logits: torch.Tensor,\n",
    "    residue_weights: Optional[torch.Tensor] = None,\n",
    "    max_bin: int = 31,\n",
    "    no_bins: int = 64,\n",
    "    eps: float = 1e-8,\n",
    "    **kwargs,\n",
    ") -> torch.Tensor:\n",
    "    if residue_weights is None:\n",
    "        residue_weights = logits.new_ones(logits.shape[-2])\n",
    "\n",
    "    boundaries = torch.linspace(\n",
    "        0, max_bin, steps=(no_bins - 1), device=logits.device\n",
    "    )\n",
    "\n",
    "    bin_centers = _calculate_bin_centers(boundaries)\n",
    "    torch.sum(residue_weights)\n",
    "    n = logits.shape[-2]\n",
    "    clipped_n = max(n, 19)\n",
    "\n",
    "    d0 = 1.24 * (clipped_n - 15) ** (1.0 / 3) - 1.8\n",
    "\n",
    "    probs = torch.nn.functional.softmax(logits, dim=-1)\n",
    "\n",
    "    tm_per_bin = 1.0 / (1 + (bin_centers ** 2) / (d0 ** 2))\n",
    "    predicted_tm_term = torch.sum(probs * tm_per_bin, dim=-1)\n",
    "\n",
    "    normed_residue_mask = residue_weights / (eps + residue_weights.sum())\n",
    "    per_alignment = torch.sum(predicted_tm_term * normed_residue_mask, dim=-1)\n",
    "\n",
    "    weighted = per_alignment * residue_weights\n",
    "     \n",
    "    argmax = (weighted == torch.max(weighted)).nonzero()[0]\n",
    "    return per_alignment[tuple(argmax)]\n",
    "\n",
    "def compute_predicted_aligned_error(\n",
    "    logits: torch.Tensor,\n",
    "    max_bin: int = 31,\n",
    "    no_bins: int = 64,\n",
    "    **kwargs,\n",
    ") -> Dict[str, torch.Tensor]:\n",
    "    \"\"\"Computes aligned confidence metrics from logits.\n",
    "\n",
    "    Args:\n",
    "      logits: [*, num_res, num_res, num_bins] the logits output from\n",
    "        PredictedAlignedErrorHead.\n",
    "      max_bin: Maximum bin value\n",
    "      no_bins: Number of bins\n",
    "    Returns:\n",
    "      aligned_confidence_probs: [*, num_res, num_res, num_bins] the predicted\n",
    "        aligned error probabilities over bins for each residue pair.\n",
    "      predicted_aligned_error: [*, num_res, num_res] the expected aligned distance\n",
    "        error for each pair of residues.\n",
    "      max_predicted_aligned_error: [*] the maximum predicted error possible.\n",
    "    \"\"\"\n",
    "    boundaries = torch.linspace(\n",
    "        0, max_bin, steps=(no_bins - 1), device=logits.device\n",
    "    )\n",
    "\n",
    "    aligned_confidence_probs = torch.nn.functional.softmax(logits, dim=-1)\n",
    "    (\n",
    "        predicted_aligned_error,\n",
    "        max_predicted_aligned_error,\n",
    "    ) = _calculate_expected_aligned_error(\n",
    "        alignment_confidence_breaks=boundaries,\n",
    "        aligned_distance_error_probs=aligned_confidence_probs,\n",
    "    )\n",
    "\n",
    "    return {\n",
    "        \"aligned_confidence_probs\": aligned_confidence_probs,\n",
    "        \"predicted_aligned_error\": predicted_aligned_error,\n",
    "        \"max_predicted_aligned_error\": max_predicted_aligned_error,\n",
    "    }\n",
    "\n",
    "import residue_constants as rc\n",
    "\n",
    "def make_atom14_masks(protein):\n",
    "    \"\"\"Construct denser atom positions (14 dimensions instead of 37).\"\"\"\n",
    "    restype_atom14_to_atom37 = []\n",
    "    restype_atom37_to_atom14 = []\n",
    "    restype_atom14_mask = []\n",
    "\n",
    "    for rt in rc.restypes:\n",
    "        atom_names = rc.restype_name_to_atom14_names[rc.restype_1to3[rt]]\n",
    "        restype_atom14_to_atom37.append(\n",
    "            [(rc.atom_order[name] if name else 0) for name in atom_names]\n",
    "        )\n",
    "        atom_name_to_idx14 = {name: i for i, name in enumerate(atom_names)}\n",
    "        restype_atom37_to_atom14.append(\n",
    "            [\n",
    "                (atom_name_to_idx14[name] if name in atom_name_to_idx14 else 0)\n",
    "                for name in rc.atom_types\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        restype_atom14_mask.append(\n",
    "            [(1.0 if name else 0.0) for name in atom_names]\n",
    "        )\n",
    "\n",
    "    # Add dummy mapping for restype 'UNK'\n",
    "    restype_atom14_to_atom37.append([0] * 14)\n",
    "    restype_atom37_to_atom14.append([0] * 37)\n",
    "    restype_atom14_mask.append([0.0] * 14)\n",
    "\n",
    "    restype_atom14_to_atom37 = torch.tensor(\n",
    "        restype_atom14_to_atom37,\n",
    "        dtype=torch.int32,\n",
    "        device=protein[\"aatype\"].device,\n",
    "    )\n",
    "    restype_atom37_to_atom14 = torch.tensor(\n",
    "        restype_atom37_to_atom14,\n",
    "        dtype=torch.int32,\n",
    "        device=protein[\"aatype\"].device,\n",
    "    )\n",
    "    restype_atom14_mask = torch.tensor(\n",
    "        restype_atom14_mask,\n",
    "        dtype=torch.float32,\n",
    "        device=protein[\"aatype\"].device,\n",
    "    )\n",
    "    protein_aatype = protein['aatype'].to(torch.long)\n",
    "\n",
    "    # create the mapping for (residx, atom14) --> atom37, i.e. an array\n",
    "    # with shape (num_res, 14) containing the atom37 indices for this protein\n",
    "    residx_atom14_to_atom37 = restype_atom14_to_atom37[protein_aatype]\n",
    "    residx_atom14_mask = restype_atom14_mask[protein_aatype]\n",
    "\n",
    "    protein[\"atom14_atom_exists\"] = residx_atom14_mask\n",
    "    protein[\"residx_atom14_to_atom37\"] = residx_atom14_to_atom37.long()\n",
    "\n",
    "    # create the gather indices for mapping back\n",
    "    residx_atom37_to_atom14 = restype_atom37_to_atom14[protein_aatype]\n",
    "    protein[\"residx_atom37_to_atom14\"] = residx_atom37_to_atom14.long()\n",
    "\n",
    "    # create the corresponding mask\n",
    "    restype_atom37_mask = torch.zeros(\n",
    "        [21, 37], dtype=torch.float32, device=protein[\"aatype\"].device\n",
    "    )\n",
    "    for restype, restype_letter in enumerate(rc.restypes):\n",
    "        restype_name = rc.restype_1to3[restype_letter]\n",
    "        atom_names = rc.residue_atoms[restype_name]\n",
    "        for atom_name in atom_names:\n",
    "            atom_type = rc.atom_order[atom_name]\n",
    "            restype_atom37_mask[restype, atom_type] = 1\n",
    "\n",
    "    residx_atom37_mask = restype_atom37_mask[protein_aatype]\n",
    "    protein[\"atom37_atom_exists\"] = residx_atom37_mask\n",
    "\n",
    "    return protein\n",
    "\n",
    "\n",
    "from esm2 import ESM2\n",
    "import re\n",
    "\n",
    "def _load_model_and_alphabet_core_v2(model_data):\n",
    "    def upgrade_state_dict(state_dict):\n",
    "        \"\"\"Removes prefixes 'model.encoder.sentence_encoder.' and 'model.encoder.'.\"\"\"\n",
    "        prefixes = [\"encoder.sentence_encoder.\", \"encoder.\"]\n",
    "        pattern = re.compile(\"^\" + \"|\".join(prefixes))\n",
    "        state_dict = {pattern.sub(\"\", name): param for name, param in state_dict.items()}\n",
    "        return state_dict\n",
    "\n",
    "    cfg = model_data[\"cfg\"][\"model\"]\n",
    "    state_dict = model_data[\"model\"]\n",
    "    state_dict = upgrade_state_dict(state_dict)\n",
    "    alphabet = Alphabet.from_architecture(\"ESM-1b\")\n",
    "    model = ESM2(\n",
    "        num_layers=cfg.encoder_layers,\n",
    "        embed_dim=cfg.encoder_embed_dim,\n",
    "        attention_heads=cfg.encoder_attention_heads,\n",
    "        alphabet=alphabet,\n",
    "        token_dropout=cfg.token_dropout,\n",
    "    )\n",
    "    return model, alphabet, state_dict\n",
    "\n",
    "# Copyright (c) Meta Platforms, Inc. and affiliates.\n",
    "#\n",
    "# This source code is licensed under the MIT license found in the\n",
    "# LICENSE file in the root directory of this source tree.\n",
    "import typing as T\n",
    "from dataclasses import dataclass\n",
    "from functools import partial\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from torch import nn\n",
    "from torch.nn import LayerNorm\n",
    "\n",
    "from data import Alphabet\n",
    "from categorical_mixture import categorical_lddt\n",
    "from trunk import FoldingTrunk, FoldingTrunkConfig\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class ESMFoldConfig:\n",
    "    trunk: T.Any = FoldingTrunkConfig\n",
    "    lddt_head_hid_dim: int = 128\n",
    "\n",
    "class ESMFold(nn.Module):\n",
    "    def __init__(self, esm, esm_dict,  esmfold_config=None, **kwargs):\n",
    "        super().__init__()\n",
    "\n",
    "        self.cfg = esmfold_config if esmfold_config else ESMFoldConfig(**kwargs)\n",
    "        cfg = self.cfg\n",
    "\n",
    "        self.distogram_bins = 64\n",
    "\n",
    "        self.esm = esm\n",
    "        self.esm_dict = esm_dict \n",
    "\n",
    "        self.esm.requires_grad_(False)\n",
    "        self.esm.half()\n",
    "\n",
    "        self.esm_feats = self.esm.embed_dim\n",
    "        self.esm_attns = self.esm.num_layers * self.esm.attention_heads\n",
    "        self.register_buffer(\"af2_to_esm\", ESMFold._af2_to_esm(self.esm_dict))\n",
    "        self.esm_s_combine = nn.Parameter(torch.zeros(self.esm.num_layers + 1))\n",
    "\n",
    "        c_s = cfg.trunk.sequence_state_dim\n",
    "        c_z = cfg.trunk.pairwise_state_dim\n",
    "\n",
    "        self.esm_s_mlp = nn.Sequential(\n",
    "            LayerNorm(self.esm_feats),\n",
    "            nn.Linear(self.esm_feats, c_s),\n",
    "            nn.ReLU(),\n",
    "            nn.Linear(c_s, c_s),\n",
    "        )\n",
    "        if cfg.use_esm_attn_map:\n",
    "            self.esm_z_mlp = nn.Sequential(\n",
    "                LayerNorm(self.esm_attns),\n",
    "                nn.Linear(self.esm_attns, c_z),\n",
    "                nn.ReLU(),\n",
    "                nn.Linear(c_z, c_z),\n",
    "            )\n",
    "\n",
    "        # 0 is padding, N is unknown residues, N + 1 is mask.\n",
    "        self.n_tokens_embed = residue_constants.restype_num + 3\n",
    "        self.pad_idx = 0\n",
    "        self.unk_idx = self.n_tokens_embed - 2\n",
    "        self.mask_idx = self.n_tokens_embed - 1\n",
    "        self.embedding = nn.Embedding(self.n_tokens_embed, c_s, padding_idx=0)\n",
    "\n",
    "        self.trunk = FoldingTrunk(**cfg.trunk)\n",
    "\n",
    "        self.distogram_head = nn.Linear(c_z, self.distogram_bins)\n",
    "        self.ptm_head = nn.Linear(c_z, self.distogram_bins)\n",
    "        self.lm_head = nn.Linear(c_s, self.n_tokens_embed)\n",
    "        self.lddt_bins = 50\n",
    "        self.lddt_head = nn.Sequential(\n",
    "            nn.LayerNorm(cfg.trunk.structure_module.c_s),\n",
    "            nn.Linear(cfg.trunk.structure_module.c_s, cfg.lddt_head_hid_dim),\n",
    "            nn.Linear(cfg.lddt_head_hid_dim, cfg.lddt_head_hid_dim),\n",
    "            nn.Linear(cfg.lddt_head_hid_dim, 37 * self.lddt_bins),\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def _af2_to_esm(d: Alphabet):\n",
    "        # Remember that t is shifted from residue_constants by 1 (0 is padding).\n",
    "        esm_reorder = [d.padding_idx] + [\n",
    "            d.get_idx(v) for v in residue_constants.restypes_with_x\n",
    "        ]\n",
    "        return torch.tensor(esm_reorder)\n",
    "\n",
    "    def _af2_idx_to_esm_idx(self, aa, mask):\n",
    "        aa = (aa + 1).masked_fill(mask != 1, 0)\n",
    "        return self.af2_to_esm[aa]\n",
    "\n",
    "    def _compute_language_model_representations(\n",
    "        self, esmaa: torch.Tensor\n",
    "    ) -> torch.Tensor:\n",
    "        \"\"\"Adds bos/eos tokens for the language model, since the structure module doesn't use these.\"\"\"\n",
    "        batch_size = esmaa.size(0)\n",
    "\n",
    "        bosi, eosi = self.esm_dict.cls_idx, self.esm_dict.eos_idx\n",
    "        bos = esmaa.new_full((batch_size, 1), bosi)\n",
    "        eos = esmaa.new_full((batch_size, 1), self.esm_dict.padding_idx)\n",
    "        esmaa = torch.cat([bos, esmaa, eos], dim=1)\n",
    "        # Use the first padding index as eos during inference.\n",
    "        esmaa[range(batch_size), (esmaa != 1).sum(1)] = eosi\n",
    "\n",
    "        res = self.esm(\n",
    "            esmaa,\n",
    "            repr_layers=range(self.esm.num_layers + 1),\n",
    "            need_head_weights=self.cfg.use_esm_attn_map,\n",
    "        )\n",
    "        esm_s = torch.stack(\n",
    "            [v for _, v in sorted(res[\"representations\"].items())], dim=2\n",
    "        )\n",
    "        esm_s = esm_s[:, 1:-1]  # B, L, nLayers, C\n",
    "        esm_z = (\n",
    "            res[\"attentions\"].permute(0, 4, 3, 1, 2).flatten(3, 4)[:, 1:-1, 1:-1, :]\n",
    "            if self.cfg.use_esm_attn_map\n",
    "            else None\n",
    "        )\n",
    "        return esm_s, esm_z\n",
    "\n",
    "    def _mask_inputs_to_esm(self, esmaa, pattern):\n",
    "        new_esmaa = esmaa.clone()\n",
    "        new_esmaa[pattern == 1] = self.esm_dict.mask_idx\n",
    "        return new_esmaa\n",
    "\n",
    "    def forward(\n",
    "        self,\n",
    "        aa: torch.Tensor,\n",
    "        mask: T.Optional[torch.Tensor] = None,\n",
    "        residx: T.Optional[torch.Tensor] = None,\n",
    "        masking_pattern: T.Optional[torch.Tensor] = None,\n",
    "        num_recycles: T.Optional[int] = None,\n",
    "    ):\n",
    "        \"\"\"Runs a forward pass given input tokens. Use `model.infer` to\n",
    "        run inference from a sequence.\n",
    "\n",
    "        Args:\n",
    "            aa (torch.Tensor): Tensor containing indices corresponding to amino acids. Indices match\n",
    "                openfold.np.residue_constants.restype_order_with_x.\n",
    "            mask (torch.Tensor): Binary tensor with 1 meaning position is unmasked and 0 meaning position is masked.\n",
    "            residx (torch.Tensor): Residue indices of amino acids. Will assume contiguous if not provided.\n",
    "            masking_pattern (torch.Tensor): Optional masking to pass to the input. Binary tensor of the same size\n",
    "                as `aa`. Positions with 1 will be masked. ESMFold sometimes produces different samples when\n",
    "                different masks are provided.\n",
    "            num_recycles (int): How many recycle iterations to perform. If None, defaults to training max\n",
    "                recycles, which is 3.\n",
    "        \"\"\"\n",
    "\n",
    "        if mask is None:\n",
    "            mask = torch.ones_like(aa)\n",
    "\n",
    "        B = aa.shape[0]\n",
    "        L = aa.shape[1]\n",
    "        device = aa.device\n",
    "\n",
    "        if residx is None:\n",
    "            residx = torch.arange(L, device=device).expand_as(aa)\n",
    "\n",
    "        # === ESM ===\n",
    "        esmaa = self._af2_idx_to_esm_idx(aa, mask)\n",
    "\n",
    "        if masking_pattern is not None:\n",
    "            esmaa = self._mask_inputs_to_esm(esmaa, masking_pattern)\n",
    "\n",
    "        esm_s, esm_z = self._compute_language_model_representations(esmaa)\n",
    "\n",
    "        # Convert esm_s to the precision used by the trunk and\n",
    "        # the structure module. These tensors may be a lower precision if, for example,\n",
    "        # we're running the language model in fp16 precision.\n",
    "        esm_s = esm_s.to(self.esm_s_combine.dtype)\n",
    "        esm_s = esm_s.detach()\n",
    "\n",
    "        # === preprocessing ===\n",
    "        esm_s = (self.esm_s_combine.softmax(0).unsqueeze(0) @ esm_s).squeeze(2)\n",
    "\n",
    "        s_s_0 = self.esm_s_mlp(esm_s)\n",
    "        if self.cfg.use_esm_attn_map:\n",
    "            esm_z = esm_z.to(self.esm_s_combine.dtype)\n",
    "            esm_z = esm_z.detach()\n",
    "            s_z_0 = self.esm_z_mlp(esm_z)\n",
    "        else:\n",
    "            s_z_0 = s_s_0.new_zeros(B, L, L, self.cfg.trunk.pairwise_state_dim)\n",
    "\n",
    "        s_s_0 += self.embedding(aa)\n",
    "\n",
    "        structure: dict = self.trunk(\n",
    "            s_s_0, s_z_0, aa, residx, mask, no_recycles=num_recycles\n",
    "        )\n",
    "        # Documenting what we expect:\n",
    "        structure = {\n",
    "            k: v\n",
    "            for k, v in structure.items()\n",
    "            if k\n",
    "            in [\n",
    "                \"s_z\",\n",
    "                \"s_s\",\n",
    "                \"frames\",\n",
    "                \"sidechain_frames\",\n",
    "                \"unnormalized_angles\",\n",
    "                \"angles\",\n",
    "                \"positions\",\n",
    "                \"states\",\n",
    "            ]\n",
    "        }\n",
    "\n",
    "        disto_logits = self.distogram_head(structure[\"s_z\"])\n",
    "        disto_logits = (disto_logits + disto_logits.transpose(1, 2)) / 2\n",
    "        structure[\"distogram_logits\"] = disto_logits\n",
    "\n",
    "        lm_logits = self.lm_head(structure[\"s_s\"])\n",
    "        structure[\"lm_logits\"] = lm_logits\n",
    "\n",
    "        structure[\"aatype\"] = aa\n",
    "        make_atom14_masks(structure)\n",
    "\n",
    "        for k in [\n",
    "            \"atom14_atom_exists\",\n",
    "            \"atom37_atom_exists\",\n",
    "        ]:\n",
    "            structure[k] *= mask.unsqueeze(-1)\n",
    "        structure[\"residue_index\"] = residx\n",
    "\n",
    "        lddt_head = self.lddt_head(structure[\"states\"]).reshape(\n",
    "            structure[\"states\"].shape[0], B, L, -1, self.lddt_bins\n",
    "        )\n",
    "        structure[\"lddt_head\"] = lddt_head\n",
    "        plddt = categorical_lddt(lddt_head[-1], bins=self.lddt_bins)\n",
    "        structure[\"plddt\"] = (\n",
    "            100 * plddt\n",
    "        )  # we predict plDDT between 0 and 1, scale to be between 0 and 100.\n",
    "\n",
    "        ptm_logits = self.ptm_head(structure[\"s_z\"])\n",
    "\n",
    "        seqlen = mask.type(torch.int64).sum(1)\n",
    "        structure[\"ptm_logits\"] = ptm_logits\n",
    "        structure[\"ptm\"] = torch.stack(\n",
    "            [\n",
    "                compute_tm(\n",
    "                    batch_ptm_logits[None, :sl, :sl],\n",
    "                    max_bins=31,\n",
    "                    no_bins=self.distogram_bins,\n",
    "                )\n",
    "                for batch_ptm_logits, sl in zip(ptm_logits, seqlen)\n",
    "            ]\n",
    "        )\n",
    "        structure.update(\n",
    "            compute_predicted_aligned_error(\n",
    "                ptm_logits, max_bin=31, no_bins=self.distogram_bins\n",
    "            )\n",
    "        )\n",
    "\n",
    "        return structure\n",
    "    \n",
    "    def set_chunk_size(self, chunk_size: T.Optional[int]):\n",
    "        # This parameter means the axial attention will be computed\n",
    "        # in a chunked manner. This should make the memory used more or less O(L) instead of O(L^2).\n",
    "        # It's equivalent to running a for loop over chunks of the dimension we're iterative over,\n",
    "        # where the chunk_size is the size of the chunks, so 128 would mean to parse 128-lengthed chunks.\n",
    "        # Setting the value to None will return to default behavior, disable chunking.\n",
    "        self.trunk.set_chunk_size(chunk_size)\n",
    "\n",
    "    @property\n",
    "    def device(self):\n",
    "        return self.esm_s_combine.device\n",
    "\n",
    "\n",
    "MODEL_NAME = \"esm2_t36_3B_UR50D\"\n",
    "\n",
    "# Load the main model data\n",
    "model_data = torch.load(f'../model/{MODEL_NAME}.pt', \n",
    "                        mmap=True, weights_only=False)\n",
    "print(\"Main model loaded\")\n",
    "\n",
    "# Get model, alphabet, and upgraded state_dict\n",
    "esm, alphabet, esm_dict = _load_model_and_alphabet_core_v2(model_data)\n",
    "\n",
    "ESM_FOLD_MODEL_NAME = \"esm_fold_v1\"\n",
    "\n",
    "esm_fold_data = torch.load(f'../model/{ESM_FOLD_MODEL_NAME}.pt', \n",
    "                        mmap=True, weights_only=False)\n",
    "\n",
    "cfg = esm_fold_data[\"cfg\"][\"model\"]\n",
    "model_state = esm_fold_data[\"model\"]\n",
    "model = ESMFold(esm, alphabet, esmfold_config=cfg)\n",
    "\n",
    "expected_keys = set(model.state_dict().keys())\n",
    "found_keys = set(model_state.keys())\n",
    "\n",
    "missing_essential_keys = []\n",
    "for missing_key in expected_keys - found_keys:\n",
    "    if not missing_key.startswith(\"esm.\"):\n",
    "        missing_essential_keys.append(missing_key)\n",
    "\n",
    "if missing_essential_keys:\n",
    "    print(\"MISSING ESSENTIAL KEYS\")\n",
    "    raise RuntimeError(f\"Keys '{', '.join(missing_essential_keys)}' are missing.\")\n",
    "\n",
    "model.load_state_dict(model_state, strict=False)\n",
    "model.to(DEVICE)\n",
    "\n",
    "# Load sequences from file\n",
    "sequences = []\n",
    "seq_names = []\n",
    "# A real protein sequence (Human Insulin)\n",
    "sequences = [\"MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\"]\n",
    "\n",
    "# AlphaFold2 vocabulary order, which the model's embedding layer expects\n",
    "AF2_VOCAB = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', 'X']\n",
    "af2_map = {amino_acid: i for i, amino_acid in enumerate(AF2_VOCAB)}\n",
    "\n",
    "# Create a map to translate from ESM's token indices to AlphaFold2's indices\n",
    "esm_to_af2_map = torch.full((len(alphabet),), fill_value=af2_map['X'], dtype=torch.long)\n",
    "for amino_acid, af2_idx in af2_map.items():\n",
    "    if amino_acid in alphabet.tok_to_idx:\n",
    "        esm_idx = alphabet.tok_to_idx[amino_acid]\n",
    "        esm_to_af2_map[esm_idx] = af2_idx\n",
    "esm_to_af2_map = esm_to_af2_map.to(DEVICE)\n",
    "\n",
    "# --- 3. Tokenize and Convert ---\n",
    "\n",
    "# Use ESM's batch converter to handle tokenization and padding\n",
    "_, _, tokens = alphabet.get_batch_converter()([( \"protein1\", sequences[0])])\n",
    "tokens = tokens.to(DEVICE)\n",
    "\n",
    "# Slice off the <cls> and <eos> tokens, as they are not part of the sequence\n",
    "sequence_tokens = tokens[:, 1:-1]\n",
    "\n",
    "# Translate from ESM token indices to the required AlphaFold2 indices\n",
    "aa = esm_to_af2_map[sequence_tokens]\n",
    "\n",
    "# Create the mask by identifying non-padding tokens\n",
    "mask = (sequence_tokens != alphabet.padding_idx).long()\n",
    "\n",
    "# --- 4. Prepare for Forward Pass ---\n",
    "\n",
    "batch_size, sequence_length = aa.shape\n",
    "residx = torch.arange(sequence_length, device=DEVICE).unsqueeze(0).expand(batch_size, -1)\n",
    "num_recycles = 3\n",
    "masking_pattern = torch.zeros(batch_size, sequence_length)  # No masking\n",
    "\n",
    "print(f\"Input 'aa' tensor shape: {aa.shape}\")\n",
    "print(f\"Input 'mask' tensor shape: {mask.shape}\")\n",
    "\n",
    "# Call the forward function\n",
    "output = model.forward(\n",
    "    aa=aa,\n",
    "    mask=mask,\n",
    "    residx=residx,\n",
    "    masking_pattern=masking_pattern,\n",
    "    num_recycles=num_recycles\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30117f3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pos37: (110, 14, 3) aa_idx: (110,) NaNs: False\n"
     ]
    },
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_17613512088053458\"  style=\"position: relative; width: 900px; height: 600px;\">\n        <p id=\"3dmolwarning_17613512088053458\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdn.jsdelivr.net/npm/3dmol@2.5.3/build/3Dmol-min.js');\n}\n\nvar viewer_17613512088053458 = null;\nvar warn = document.getElementById(\"3dmolwarning_17613512088053458\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_17613512088053458 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17613512088053458\"),{backgroundColor:\"white\"});\nviewer_17613512088053458.zoomTo();\n\tviewer_17613512088053458.addModel(\"ATOM      1   N MET A   1      -2.291 -14.445 -41.302  1.00  0.00           N\\nATOM      2 CA  MET A   1      -2.968 -13.376 -40.574  1.00  0.00           C\\nATOM      3   C MET A   1      -3.255 -13.795 -39.136  1.00  0.00           C\\nATOM      4   O MET A   1      -3.010 -13.029 -38.202  1.00  0.00           O\\nATOM      5   N ALA A   2      -3.712 -15.067 -38.982  1.00  0.00           N\\nATOM      6 CA  ALA A   2      -4.057 -15.557 -37.651  1.00  0.00           C\\nATOM      7   C ALA A   2      -2.824 -15.626 -36.754  1.00  0.00           C\\nATOM      8   O ALA A   2      -2.886 -15.272 -35.574  1.00  0.00           O\\nATOM      9   N LEU A   3      -1.670 -15.930 -37.250  1.00  0.00           N\\nATOM     10 CA  LEU A   3      -0.412 -16.029 -36.519  1.00  0.00           C\\nATOM     11   C LEU A   3       0.055 -14.655 -36.051  1.00  0.00           C\\nATOM     12   O LEU A   3       0.477 -14.496 -34.903  1.00  0.00           O\\nATOM     13   N TRP A   4      -0.036 -13.714 -36.947  1.00  0.00           N\\nATOM     14 CA  TRP A   4       0.364 -12.346 -36.635  1.00  0.00           C\\nATOM     15   C TRP A   4      -0.495 -11.769 -35.515  1.00  0.00           C\\nATOM     16   O TRP A   4       0.024 -11.147 -34.584  1.00  0.00           O\\nATOM     17   N MET A   5      -1.765 -11.945 -35.634  1.00  0.00           N\\nATOM     18 CA  MET A   5      -2.701 -11.453 -34.627  1.00  0.00           C\\nATOM     19   C MET A   5      -2.411 -12.075 -33.265  1.00  0.00           C\\nATOM     20   O MET A   5      -2.414 -11.380 -32.247  1.00  0.00           O\\nATOM     21   N ARG A   6      -2.081 -13.374 -33.270  1.00  0.00           N\\nATOM     22 CA  ARG A   6      -1.714 -14.095 -32.056  1.00  0.00           C\\nATOM     23   C ARG A   6      -0.435 -13.530 -31.448  1.00  0.00           C\\nATOM     24   O ARG A   6      -0.355 -13.326 -30.235  1.00  0.00           O\\nATOM     25   N LEU A   7       0.480 -13.328 -32.248  1.00  0.00           N\\nATOM     26 CA  LEU A   7       1.756 -12.767 -31.818  1.00  0.00           C\\nATOM     27   C LEU A   7       1.565 -11.376 -31.222  1.00  0.00           C\\nATOM     28   O LEU A   7       2.123 -11.065 -30.167  1.00  0.00           O\\nATOM     29   N LEU A   8       0.760 -10.583 -31.831  1.00  0.00           N\\nATOM     30 CA  LEU A   8       0.481  -9.226 -31.372  1.00  0.00           C\\nATOM     31   C LEU A   8      -0.221  -9.244 -30.019  1.00  0.00           C\\nATOM     32   O LEU A   8       0.138  -8.483 -29.118  1.00  0.00           O\\nATOM     33   N PRO A   9      -1.223 -10.139 -29.816  1.00  0.00           N\\nATOM     34 CA  PRO A   9      -1.909 -10.285 -28.530  1.00  0.00           C\\nATOM     35   C PRO A   9      -0.971 -10.727 -27.409  1.00  0.00           C\\nATOM     36   O PRO A   9      -1.044 -10.202 -26.295  1.00  0.00           O\\nATOM     37   N LEU A  10      -0.100 -11.685 -27.713  1.00  0.00           N\\nATOM     38 CA  LEU A  10       0.893 -12.153 -26.752  1.00  0.00           C\\nATOM     39   C LEU A  10       1.821 -11.018 -26.334  1.00  0.00           C\\nATOM     40   O LEU A  10       2.089 -10.836 -25.144  1.00  0.00           O\\nATOM     41   N LEU A  11       2.251 -10.277 -27.318  1.00  0.00           N\\nATOM     42 CA  LEU A  11       3.143  -9.152 -27.055  1.00  0.00           C\\nATOM     43   C LEU A  11       2.455  -8.105 -26.186  1.00  0.00           C\\nATOM     44   O LEU A  11       3.039  -7.618 -25.215  1.00  0.00           O\\nATOM     45   N ALA A  12       1.297  -7.758 -26.530  1.00  0.00           N\\nATOM     46 CA  ALA A  12       0.523  -6.775 -25.776  1.00  0.00           C\\nATOM     47   C ALA A  12       0.300  -7.237 -24.339  1.00  0.00           C\\nATOM     48   O ALA A  12       0.470  -6.458 -23.397  1.00  0.00           O\\nATOM     49   N LEU A  13      -0.003  -8.503 -24.126  1.00  0.00           N\\nATOM     50 CA  LEU A  13      -0.230  -9.077 -22.804  1.00  0.00           C\\nATOM     51   C LEU A  13       1.054  -9.071 -21.981  1.00  0.00           C\\nATOM     52   O LEU A  13       1.036  -8.722 -20.798  1.00  0.00           O\\nATOM     53   N LEU A  14       2.071  -9.472 -22.642  1.00  0.00           N\\nATOM     54 CA  LEU A  14       3.378  -9.443 -21.995  1.00  0.00           C\\nATOM     55   C LEU A  14       3.734  -8.028 -21.551  1.00  0.00           C\\nATOM     56   O LEU A  14       4.184  -7.822 -20.422  1.00  0.00           O\\nATOM     57   N ALA A  15       3.527  -7.083 -22.407  1.00  0.00           N\\nATOM     58 CA  ALA A  15       3.797  -5.681 -22.098  1.00  0.00           C\\nATOM     59   C ALA A  15       2.952  -5.208 -20.919  1.00  0.00           C\\nATOM     60   O ALA A  15       3.462  -4.559 -20.002  1.00  0.00           O\\nATOM     61   N LEU A  16       1.648  -5.528 -20.870  1.00  0.00           N\\nATOM     62 CA  LEU A  16       0.747  -5.163 -19.783  1.00  0.00           C\\nATOM     63   C LEU A  16       1.215  -5.766 -18.463  1.00  0.00           C\\nATOM     64   O LEU A  16       1.229  -5.086 -17.434  1.00  0.00           O\\nATOM     65   N TRP A  17       1.663  -6.991 -18.518  1.00  0.00           N\\nATOM     66 CA  TRP A  17       2.203  -7.688 -17.355  1.00  0.00           C\\nATOM     67   C TRP A  17       3.441  -6.977 -16.820  1.00  0.00           C\\nATOM     68   O TRP A  17       3.562  -6.752 -15.613  1.00  0.00           O\\nATOM     69   N GLY A  18       4.367  -6.726 -17.673  1.00  0.00           N\\nATOM     70 CA  GLY A  18       5.560  -5.988 -17.292  1.00  0.00           C\\nATOM     71   C GLY A  18       5.253  -4.652 -16.641  1.00  0.00           C\\nATOM     72   O GLY A  18       5.836  -4.310 -15.610  1.00  0.00           O\\nATOM     73   N PRO A  19       4.364  -3.982 -17.266  1.00  0.00           N\\nATOM     74 CA  PRO A  19       3.963  -2.687 -16.711  1.00  0.00           C\\nATOM     75   C PRO A  19       3.336  -2.808 -15.324  1.00  0.00           C\\nATOM     76   O PRO A  19       3.680  -2.045 -14.418  1.00  0.00           O\\nATOM     77   N ASP A  20       2.440  -3.795 -15.088  1.00  0.00           N\\nATOM     78 CA  ASP A  20       1.799  -4.027 -13.797  1.00  0.00           C\\nATOM     79   C ASP A  20       2.828  -4.399 -12.732  1.00  0.00           C\\nATOM     80   O ASP A  20       2.778  -3.892 -11.609  1.00  0.00           O\\nATOM     81   N PRO A  21       3.795  -5.267 -13.148  1.00  0.00           N\\nATOM     82 CA  PRO A  21       4.894  -5.611 -12.244  1.00  0.00           C\\nATOM     83   C PRO A  21       5.720  -4.396 -11.829  1.00  0.00           C\\nATOM     84   O PRO A  21       6.021  -4.224 -10.645  1.00  0.00           O\\nATOM     85   N ALA A  22       6.045  -3.550 -12.828  1.00  0.00           N\\nATOM     86 CA  ALA A  22       6.843  -2.354 -12.566  1.00  0.00           C\\nATOM     87   C ALA A  22       6.104  -1.395 -11.637  1.00  0.00           C\\nATOM     88   O ALA A  22       6.674  -0.905 -10.659  1.00  0.00           O\\nATOM     89   N ALA A  23       4.887  -1.130 -11.959  1.00  0.00           N\\nATOM     90 CA  ALA A  23       4.075  -0.226 -11.149  1.00  0.00           C\\nATOM     91   C ALA A  23       3.962  -0.729  -9.713  1.00  0.00           C\\nATOM     92   O ALA A  23       4.121   0.043  -8.764  1.00  0.00           O\\nATOM     93   N ALA A  24       3.762  -1.991  -9.491  1.00  0.00           N\\nATOM     94 CA  ALA A  24       3.658  -2.611  -8.173  1.00  0.00           C\\nATOM     95   C ALA A  24       4.970  -2.488  -7.404  1.00  0.00           C\\nATOM     96   O ALA A  24       4.973  -2.140  -6.221  1.00  0.00           O\\nATOM     97   N PHE A  25       6.027  -2.842  -8.092  1.00  0.00           N\\nATOM     98 CA  PHE A  25       7.356  -2.692  -7.510  1.00  0.00           C\\nATOM     99   C PHE A  25       7.579  -1.263  -7.031  1.00  0.00           C\\nATOM    100   O PHE A  25       8.021  -1.043  -5.901  1.00  0.00           O\\nATOM    101   N VAL A  26       7.221  -0.278  -7.839  1.00  0.00           N\\nATOM    102 CA  VAL A  26       7.403   1.134  -7.518  1.00  0.00           C\\nATOM    103   C VAL A  26       6.556   1.501  -6.301  1.00  0.00           C\\nATOM    104   O VAL A  26       7.045   2.137  -5.365  1.00  0.00           O\\nATOM    105   N ASN A  27       5.273   1.140  -6.310  1.00  0.00           N\\nATOM    106 CA  ASN A  27       4.374   1.420  -5.195  1.00  0.00           C\\nATOM    107   C ASN A  27       4.889   0.812  -3.894  1.00  0.00           C\\nATOM    108   O ASN A  27       4.868   1.463  -2.848  1.00  0.00           O\\nATOM    109   N GLN A  28       5.411  -0.399  -3.977  1.00  0.00           N\\nATOM    110 CA  GLN A  28       5.980  -1.105  -2.834  1.00  0.00           C\\nATOM    111   C GLN A  28       7.199  -0.370  -2.284  1.00  0.00           C\\nATOM    112   O GLN A  28       7.334  -0.203  -1.070  1.00  0.00           O\\nATOM    113   N HIS A  29       8.128  -0.030  -3.200  1.00  0.00           N\\nATOM    114 CA  HIS A  29       9.320   0.725  -2.831  1.00  0.00           C\\nATOM    115   C HIS A  29       8.951   2.026  -2.127  1.00  0.00           C\\nATOM    116   O HIS A  29       9.526   2.360  -1.088  1.00  0.00           O\\nATOM    117   N LEU A  30       7.976   2.763  -2.637  1.00  0.00           N\\nATOM    118 CA  LEU A  30       7.540   4.036  -2.074  1.00  0.00           C\\nATOM    119   C LEU A  30       6.967   3.843  -0.674  1.00  0.00           C\\nATOM    120   O LEU A  30       7.316   4.579   0.252  1.00  0.00           O\\nATOM    121   N CYS A  31       6.071   2.774  -0.424  1.00  0.00           N\\nATOM    122 CA  CYS A  31       5.470   2.468   0.869  1.00  0.00           C\\nATOM    123   C CYS A  31       6.538   2.111   1.896  1.00  0.00           C\\nATOM    124   O CYS A  31       6.503   2.595   3.028  1.00  0.00           O\\nATOM    125   N GLY A  32       7.541   1.231   1.426  1.00  0.00           N\\nATOM    126 CA  GLY A  32       8.670   0.909   2.284  1.00  0.00           C\\nATOM    127   C GLY A  32       9.466   2.129   2.706  1.00  0.00           C\\nATOM    128   O GLY A  32       9.792   2.287   3.884  1.00  0.00           O\\nATOM    129   N SER A  33       9.755   3.026   1.763  1.00  0.00           N\\nATOM    130 CA  SER A  33      10.520   4.239   2.032  1.00  0.00           C\\nATOM    131   C SER A  33       9.796   5.141   3.026  1.00  0.00           C\\nATOM    132   O SER A  33      10.400   5.632   3.981  1.00  0.00           O\\nATOM    133   N HIS A  34       8.517   5.416   2.789  1.00  0.00           N\\nATOM    134 CA  HIS A  34       7.711   6.255   3.668  1.00  0.00           C\\nATOM    135   C HIS A  34       7.686   5.702   5.089  1.00  0.00           C\\nATOM    136   O HIS A  34       7.826   6.455   6.055  1.00  0.00           O\\nATOM    137   N LEU A  35       7.637   4.366   5.194  1.00  0.00           N\\nATOM    138 CA  LEU A  35       7.601   3.720   6.502  1.00  0.00           C\\nATOM    139   C LEU A  35       8.935   3.877   7.223  1.00  0.00           C\\nATOM    140   O LEU A  35       8.968   4.202   8.413  1.00  0.00           O\\nATOM    141   N VAL A  36       9.998   3.641   6.490  1.00  0.00           N\\nATOM    142 CA  VAL A  36      11.341   3.820   7.031  1.00  0.00           C\\nATOM    143   C VAL A  36      11.495   5.239   7.573  1.00  0.00           C\\nATOM    144   O VAL A  36      11.984   5.435   8.688  1.00  0.00           O\\nATOM    145   N GLU A  37      11.034   6.194   6.789  1.00  0.00           N\\nATOM    146 CA  GLU A  37      11.116   7.595   7.191  1.00  0.00           C\\nATOM    147   C GLU A  37      10.302   7.854   8.456  1.00  0.00           C\\nATOM    148   O GLU A  37      10.783   8.502   9.388  1.00  0.00           O\\nATOM    149   N ALA A  38       9.074   7.347   8.500  1.00  0.00           N\\nATOM    150 CA  ALA A  38       8.211   7.508   9.667  1.00  0.00           C\\nATOM    151   C ALA A  38       8.850   6.897  10.911  1.00  0.00           C\\nATOM    152   O ALA A  38       8.836   7.503  11.985  1.00  0.00           O\\nATOM    153   N LEU A  39       9.461   5.702  10.755  1.00  0.00           N\\nATOM    154 CA  LEU A  39      10.163   5.023  11.839  1.00  0.00           C\\nATOM    155   C LEU A  39      11.324   5.871  12.349  1.00  0.00           C\\nATOM    156   O LEU A  39      11.492   6.037  13.559  1.00  0.00           O\\nATOM    157   N TYR A  40      12.132   6.392  11.416  1.00  0.00           N\\nATOM    158 CA  TYR A  40      13.263   7.239  11.777  1.00  0.00           C\\nATOM    159   C TYR A  40      12.801   8.465  12.555  1.00  0.00           C\\nATOM    160   O TYR A  40      13.380   8.806  13.589  1.00  0.00           O\\nATOM    161   N LEU A  41      11.712   9.114  12.128  1.00  0.00           N\\nATOM    162 CA  LEU A  41      11.185  10.310  12.776  1.00  0.00           C\\nATOM    163   C LEU A  41      10.718   9.998  14.194  1.00  0.00           C\\nATOM    164   O LEU A  41      11.036  10.732  15.132  1.00  0.00           O\\nATOM    165   N VAL A  42      10.022   8.859  14.375  1.00  0.00           N\\nATOM    166 CA  VAL A  42       9.500   8.463  15.678  1.00  0.00           C\\nATOM    167   C VAL A  42      10.656   8.173  16.632  1.00  0.00           C\\nATOM    168   O VAL A  42      10.626   8.578  17.796  1.00  0.00           O\\nATOM    169   N CYS A  43      11.669   7.462  16.099  1.00  0.00           N\\nATOM    170 CA  CYS A  43      12.864   7.175  16.885  1.00  0.00           C\\nATOM    171   C CYS A  43      13.544   8.463  17.333  1.00  0.00           C\\nATOM    172   O CYS A  43      13.907   8.603  18.503  1.00  0.00           O\\nATOM    173   N GLY A  44      13.711   9.430  16.400  1.00  0.00           N\\nATOM    174 CA  GLY A  44      14.318  10.706  16.744  1.00  0.00           C\\nATOM    175   C GLY A  44      13.535  11.477  17.790  1.00  0.00           C\\nATOM    176   O GLY A  44      14.113  11.996  18.748  1.00  0.00           O\\nATOM    177   N GLU A  45      12.221  11.581  17.581  1.00  0.00           N\\nATOM    178 CA  GLU A  45      11.365  12.295  18.524  1.00  0.00           C\\nATOM    179   C GLU A  45      11.451  11.683  19.919  1.00  0.00           C\\nATOM    180   O GLU A  45      11.524  12.405  20.916  1.00  0.00           O\\nATOM    181   N ARG A  46      11.588  10.390  19.961  1.00  0.00           N\\nATOM    182 CA  ARG A  46      11.727   9.676  21.226  1.00  0.00           C\\nATOM    183   C ARG A  46      13.059  10.000  21.895  1.00  0.00           C\\nATOM    184   O ARG A  46      13.113  10.223  23.106  1.00  0.00           O\\nATOM    185   N GLY A  47      14.107   9.893  21.074  1.00  0.00           N\\nATOM    186 CA  GLY A  47      15.407  10.282  21.598  1.00  0.00           C\\nATOM    187   C GLY A  47      15.421  11.683  22.179  1.00  0.00           C\\nATOM    188   O GLY A  47      15.950  11.903  23.271  1.00  0.00           O\\nATOM    189   N PHE A  48      14.808  12.688  21.455  1.00  0.00           N\\nATOM    190 CA  PHE A  48      14.762  14.067  21.925  1.00  0.00           C\\nATOM    191   C PHE A  48      14.000  14.166  23.241  1.00  0.00           C\\nATOM    192   O PHE A  48      14.451  14.829  24.177  1.00  0.00           O\\nATOM    193   N PHE A  49      12.816  13.449  23.425  1.00  0.00           N\\nATOM    194 CA  PHE A  49      12.005  13.460  24.637  1.00  0.00           C\\nATOM    195   C PHE A  49      12.784  12.886  25.815  1.00  0.00           C\\nATOM    196   O PHE A  49      12.740  13.431  26.920  1.00  0.00           O\\nATOM    197   N TYR A  50      13.522  11.760  25.492  1.00  0.00           N\\nATOM    198 CA  TYR A  50      14.402  11.156  26.487  1.00  0.00           C\\nATOM    199   C TYR A  50      15.404  12.174  27.018  1.00  0.00           C\\nATOM    200   O TYR A  50      15.616  12.271  28.229  1.00  0.00           O\\nATOM    201   N THR A  51      16.073  12.781  26.080  1.00  0.00           N\\nATOM    202 CA  THR A  51      17.091  13.740  26.493  1.00  0.00           C\\nATOM    203   C THR A  51      16.486  14.822  27.384  1.00  0.00           C\\nATOM    204   O THR A  51      17.083  15.211  28.389  1.00  0.00           O\\nATOM    205   N PRO A  52      15.184  15.129  27.036  1.00  0.00           N\\nATOM    206 CA  PRO A  52      14.560  16.214  27.797  1.00  0.00           C\\nATOM    207   C PRO A  52      14.018  15.751  29.148  1.00  0.00           C\\nATOM    208   O PRO A  52      14.094  16.489  30.133  1.00  0.00           O\\nATOM    209   N LYS A  53      13.486  14.345  29.353  1.00  0.00           N\\nATOM    210 CA  LYS A  53      12.801  13.884  30.557  1.00  0.00           C\\nATOM    211   C LYS A  53      13.772  13.200  31.515  1.00  0.00           C\\nATOM    212   O LYS A  53      13.501  13.095  32.713  1.00  0.00           O\\nATOM    213   N THR A  54      14.952  12.698  31.095  1.00  0.00           N\\nATOM    214 CA  THR A  54      16.048  12.218  31.930  1.00  0.00           C\\nATOM    215   C THR A  54      16.214  13.102  33.163  1.00  0.00           C\\nATOM    216   O THR A  54      16.613  12.623  34.227  1.00  0.00           O\\nATOM    217   N ARG A  55      15.375  14.042  33.280  1.00  0.00           N\\nATOM    218 CA  ARG A  55      15.565  14.890  34.452  1.00  0.00           C\\nATOM    219   C ARG A  55      14.581  14.525  35.557  1.00  0.00           C\\nATOM    220   O ARG A  55      14.903  14.632  36.743  1.00  0.00           O\\nATOM    221   N ARG A  56      13.737  13.378  35.456  1.00  0.00           N\\nATOM    222 CA  ARG A  56      12.809  13.162  36.561  1.00  0.00           C\\nATOM    223   C ARG A  56      12.154  11.788  36.466  1.00  0.00           C\\nATOM    224   O ARG A  56      11.640  11.411  35.411  1.00  0.00           O\\nATOM    225   N GLU A  57      12.686  10.766  36.986  1.00  0.00           N\\nATOM    226 CA  GLU A  57      12.189   9.461  37.411  1.00  0.00           C\\nATOM    227   C GLU A  57      12.541   8.379  36.394  1.00  0.00           C\\nATOM    228   O GLU A  57      11.980   8.346  35.296  1.00  0.00           O\\nATOM    229   N ALA A  58      13.771   8.025  36.192  1.00  0.00           N\\nATOM    230 CA  ALA A  58      14.596   7.225  35.291  1.00  0.00           C\\nATOM    231   C ALA A  58      13.975   5.851  35.055  1.00  0.00           C\\nATOM    232   O ALA A  58      13.959   5.355  33.926  1.00  0.00           O\\nATOM    233   N GLU A  59      13.284   5.324  36.030  1.00  0.00           N\\nATOM    234 CA  GLU A  59      12.908   3.917  35.930  1.00  0.00           C\\nATOM    235   C GLU A  59      11.584   3.750  35.190  1.00  0.00           C\\nATOM    236   O GLU A  59      11.475   2.925  34.281  1.00  0.00           O\\nATOM    237   N ASP A  60      10.442   4.398  35.561  1.00  0.00           N\\nATOM    238 CA  ASP A  60       9.159   4.333  34.869  1.00  0.00           C\\nATOM    239   C ASP A  60       9.310   4.702  33.395  1.00  0.00           C\\nATOM    240   O ASP A  60       8.690   4.083  32.527  1.00  0.00           O\\nATOM    241   N LEU A  61      10.212   5.430  33.206  1.00  0.00           N\\nATOM    242 CA  LEU A  61      10.460   5.928  31.857  1.00  0.00           C\\nATOM    243   C LEU A  61      11.108   4.852  30.991  1.00  0.00           C\\nATOM    244   O LEU A  61      10.736   4.678  29.828  1.00  0.00           O\\nATOM    245   N GLN A  62      12.024   4.113  31.525  1.00  0.00           N\\nATOM    246 CA  GLN A  62      12.672   2.986  30.862  1.00  0.00           C\\nATOM    247   C GLN A  62      11.648   1.946  30.418  1.00  0.00           C\\nATOM    248   O GLN A  62      11.694   1.466  29.284  1.00  0.00           O\\nATOM    249   N VAL A  63      10.735   1.656  31.327  1.00  0.00           N\\nATOM    250 CA  VAL A  63       9.694   0.675  31.037  1.00  0.00           C\\nATOM    251   C VAL A  63       8.821   1.172  29.887  1.00  0.00           C\\nATOM    252   O VAL A  63       8.515   0.417  28.961  1.00  0.00           O\\nATOM    253   N GLY A  64       8.385   2.421  29.917  1.00  0.00           N\\nATOM    254 CA  GLY A  64       7.606   2.997  28.833  1.00  0.00           C\\nATOM    255   C GLY A  64       8.330   2.975  27.500  1.00  0.00           C\\nATOM    256   O GLY A  64       7.728   2.679  26.465  1.00  0.00           O\\nATOM    257   N GLN A  65       9.598   3.204  27.499  1.00  0.00           N\\nATOM    258 CA  GLN A  65      10.452   3.144  26.318  1.00  0.00           C\\nATOM    259   C GLN A  65      10.489   1.734  25.736  1.00  0.00           C\\nATOM    260   O GLN A  65      10.363   1.553  24.523  1.00  0.00           O\\nATOM    261   N VAL A  66      10.730   0.853  26.582  1.00  0.00           N\\nATOM    262 CA  VAL A  66      10.774  -0.549  26.181  1.00  0.00           C\\nATOM    263   C VAL A  66       9.450  -0.942  25.530  1.00  0.00           C\\nATOM    264   O VAL A  66       9.435  -1.577  24.473  1.00  0.00           O\\nATOM    265   N GLU A  67       8.387  -0.510  26.073  1.00  0.00           N\\nATOM    266 CA  GLU A  67       7.061  -0.854  25.568  1.00  0.00           C\\nATOM    267   C GLU A  67       6.807  -0.217  24.205  1.00  0.00           C\\nATOM    268   O GLU A  67       6.307  -0.875  23.290  1.00  0.00           O\\nATOM    269   N LEU A  68       7.155   1.030  24.073  1.00  0.00           N\\nATOM    270 CA  LEU A  68       6.966   1.752  22.819  1.00  0.00           C\\nATOM    271   C LEU A  68       7.820   1.147  21.710  1.00  0.00           C\\nATOM    272   O LEU A  68       7.351   0.978  20.582  1.00  0.00           O\\nATOM    273   N GLY A  69       9.144   0.915  22.015  1.00  0.00           N\\nATOM    274 CA  GLY A  69      10.011   0.200  21.092  1.00  0.00           C\\nATOM    275   C GLY A  69       9.447  -1.140  20.660  1.00  0.00           C\\nATOM    276   O GLY A  69       9.462  -1.473  19.473  1.00  0.00           O\\nATOM    277   N GLY A  70       8.946  -1.902  21.656  1.00  0.00           N\\nATOM    278 CA  GLY A  70       8.326  -3.180  21.343  1.00  0.00           C\\nATOM    279   C GLY A  70       7.128  -3.052  20.422  1.00  0.00           C\\nATOM    280   O GLY A  70       6.999  -3.805  19.454  1.00  0.00           O\\nATOM    281   N GLY A  71       6.206  -2.243  20.712  1.00  0.00           N\\nATOM    282 CA  GLY A  71       5.058  -1.993  19.855  1.00  0.00           C\\nATOM    283   C GLY A  71       5.442  -1.629  18.433  1.00  0.00           C\\nATOM    284   O GLY A  71       4.850  -2.133  17.477  1.00  0.00           O\\nATOM    285   N PRO A  72       6.390  -0.768  18.246  1.00  0.00           N\\nATOM    286 CA  PRO A  72       6.884  -0.400  16.917  1.00  0.00           C\\nATOM    287   C PRO A  72       7.444  -1.593  16.145  1.00  0.00           C\\nATOM    288   O PRO A  72       7.153  -1.756  14.957  1.00  0.00           O\\nATOM    289   N GLY A  73       8.363  -2.260  16.892  1.00  0.00           N\\nATOM    290 CA  GLY A  73       8.868  -3.481  16.286  1.00  0.00           C\\nATOM    291   C GLY A  73       7.767  -4.410  15.810  1.00  0.00           C\\nATOM    292   O GLY A  73       7.819  -4.920  14.689  1.00  0.00           O\\nATOM    293   N ALA A  74       6.816  -4.578  16.633  1.00  0.00           N\\nATOM    294 CA  ALA A  74       5.692  -5.447  16.292  1.00  0.00           C\\nATOM    295   C ALA A  74       4.946  -4.922  15.068  1.00  0.00           C\\nATOM    296   O ALA A  74       4.627  -5.685  14.154  1.00  0.00           O\\nATOM    297   N GLY A  75       4.575  -3.587  14.998  1.00  0.00           N\\nATOM    298 CA  GLY A  75       3.943  -2.975  13.840  1.00  0.00           C\\nATOM    299   C GLY A  75       4.737  -3.156  12.560  1.00  0.00           C\\nATOM    300   O GLY A  75       4.173  -3.482  11.514  1.00  0.00           O\\nATOM    301   N SER A  76       6.093  -3.023  12.656  1.00  0.00           N\\nATOM    302 CA  SER A  76       6.984  -3.243  11.521  1.00  0.00           C\\nATOM    303   C SER A  76       6.896  -4.681  11.019  1.00  0.00           C\\nATOM    304   O SER A  76       6.797  -4.918   9.814  1.00  0.00           O\\nATOM    305   N LEU A  77       6.971  -5.592  11.997  1.00  0.00           N\\nATOM    306 CA  LEU A  77       6.894  -7.008  11.656  1.00  0.00           C\\nATOM    307   C LEU A  77       5.576  -7.329  10.959  1.00  0.00           C\\nATOM    308   O LEU A  77       5.562  -8.018   9.936  1.00  0.00           O\\nATOM    309   N GLN A  78       4.433  -6.891  11.513  1.00  0.00           N\\nATOM    310 CA  GLN A  78       3.108  -7.127  10.948  1.00  0.00           C\\nATOM    311   C GLN A  78       3.011  -6.580   9.527  1.00  0.00           C\\nATOM    312   O GLN A  78       2.496  -7.253   8.631  1.00  0.00           O\\nATOM    313   N PRO A  79       3.546  -5.435   9.267  1.00  0.00           N\\nATOM    314 CA  PRO A  79       3.511  -4.840   7.929  1.00  0.00           C\\nATOM    315   C PRO A  79       4.351  -5.616   6.916  1.00  0.00           C\\nATOM    316   O PRO A  79       3.908  -5.843   5.787  1.00  0.00           O\\nATOM    317   N LEU A  80       5.572  -5.953   7.372  1.00  0.00           N\\nATOM    318 CA  LEU A  80       6.422  -6.803   6.545  1.00  0.00           C\\nATOM    319   C LEU A  80       5.689  -8.079   6.145  1.00  0.00           C\\nATOM    320   O LEU A  80       5.728  -8.484   4.981  1.00  0.00           O\\nATOM    321   N ALA A  81       4.986  -8.697   7.140  1.00  0.00           N\\nATOM    322 CA  ALA A  81       4.232  -9.922   6.888  1.00  0.00           C\\nATOM    323   C ALA A  81       3.116  -9.681   5.875  1.00  0.00           C\\nATOM    324   O ALA A  81       2.940 -10.463   4.938  1.00  0.00           O\\nATOM    325   N LEU A  82       2.387  -8.605   6.024  1.00  0.00           N\\nATOM    326 CA  LEU A  82       1.311  -8.232   5.113  1.00  0.00           C\\nATOM    327   C LEU A  82       1.843  -8.028   3.698  1.00  0.00           C\\nATOM    328   O LEU A  82       1.249  -8.512   2.731  1.00  0.00           O\\nATOM    329   N GLU A  83       3.004  -7.406   3.571  1.00  0.00           N\\nATOM    330 CA  GLU A  83       3.668  -7.167   2.293  1.00  0.00           C\\nATOM    331   C GLU A  83       4.048  -8.480   1.615  1.00  0.00           C\\nATOM    332   O GLU A  83       3.810  -8.659   0.419  1.00  0.00           O\\nATOM    333   N GLY A  84       4.736  -9.356   2.399  1.00  0.00           N\\nATOM    334 CA  GLY A  84       5.068 -10.676   1.885  1.00  0.00           C\\nATOM    335   C GLY A  84       3.861 -11.436   1.370  1.00  0.00           C\\nATOM    336   O GLY A  84       3.905 -12.019   0.284  1.00  0.00           O\\nATOM    337   N SER A  85       2.763 -11.409   2.108  1.00  0.00           N\\nATOM    338 CA  SER A  85       1.534 -12.103   1.736  1.00  0.00           C\\nATOM    339   C SER A  85       0.948 -11.537   0.447  1.00  0.00           C\\nATOM    340   O SER A  85       0.566 -12.290  -0.450  1.00  0.00           O\\nATOM    341   N LEU A  86       0.928 -10.184   0.255  1.00  0.00           N\\nATOM    342 CA  LEU A  86       0.423  -9.512  -0.937  1.00  0.00           C\\nATOM    343   C LEU A  86       1.251  -9.882  -2.163  1.00  0.00           C\\nATOM    344   O LEU A  86       0.699 -10.163  -3.229  1.00  0.00           O\\nATOM    345   N GLN A  87       2.645  -9.925  -2.028  1.00  0.00           N\\nATOM    346 CA  GLN A  87       3.545 -10.342  -3.098  1.00  0.00           C\\nATOM    347   C GLN A  87       3.241 -11.768  -3.549  1.00  0.00           C\\nATOM    348   O GLN A  87       3.156 -12.040  -4.748  1.00  0.00           O\\nATOM    349   N LYS A  88       3.087 -12.683  -2.535  1.00  0.00           N\\nATOM    350 CA  LYS A  88       2.812 -14.084  -2.842  1.00  0.00           C\\nATOM    351   C LYS A  88       1.505 -14.231  -3.616  1.00  0.00           C\\nATOM    352   O LYS A  88       1.450 -14.942  -4.622  1.00  0.00           O\\nATOM    353   N ARG A  89       0.396 -13.590  -3.183  1.00  0.00           N\\nATOM    354 CA  ARG A  89      -0.914 -13.645  -3.824  1.00  0.00           C\\nATOM    355   C ARG A  89      -0.843 -13.139  -5.261  1.00  0.00           C\\nATOM    356   O ARG A  89      -1.398 -13.757  -6.171  1.00  0.00           O\\nATOM    357   N GLY A  90      -0.100 -12.084  -5.475  1.00  0.00           N\\nATOM    358 CA  GLY A  90       0.060 -11.542  -6.814  1.00  0.00           C\\nATOM    359   C GLY A  90       0.796 -12.479  -7.753  1.00  0.00           C\\nATOM    360   O GLY A  90       0.383 -12.669  -8.899  1.00  0.00           O\\nATOM    361   N ILE A  91       1.941 -13.038  -7.220  1.00  0.00           N\\nATOM    362 CA  ILE A  91       2.699 -14.013  -7.996  1.00  0.00           C\\nATOM    363   C ILE A  91       1.783 -15.156  -8.425  1.00  0.00           C\\nATOM    364   O ILE A  91       1.817 -15.588  -9.579  1.00  0.00           O\\nATOM    365   N VAL A  92       0.903 -15.641  -7.527  1.00  0.00           N\\nATOM    366 CA  VAL A  92      -0.018 -16.741  -7.798  1.00  0.00           C\\nATOM    367   C VAL A  92      -1.019 -16.324  -8.872  1.00  0.00           C\\nATOM    368   O VAL A  92      -1.263 -17.068  -9.825  1.00  0.00           O\\nATOM    369   N GLU A  93      -1.605 -15.091  -8.735  1.00  0.00           N\\nATOM    370 CA  GLU A  93      -2.576 -14.576  -9.695  1.00  0.00           C\\nATOM    371   C GLU A  93      -1.961 -14.442 -11.085  1.00  0.00           C\\nATOM    372   O GLU A  93      -2.584 -14.808 -12.084  1.00  0.00           O\\nATOM    373   N GLN A  94      -0.683 -14.027 -11.125  1.00  0.00           N\\nATOM    374 CA  GLN A  94       0.067 -13.903 -12.370  1.00  0.00           C\\nATOM    375   C GLN A  94       0.263 -15.263 -13.033  1.00  0.00           C\\nATOM    376   O GLN A  94       0.063 -15.406 -14.241  1.00  0.00           O\\nATOM    377   N CYS A  95       0.732 -16.205 -12.268  1.00  0.00           N\\nATOM    378 CA  CYS A  95       0.941 -17.557 -12.774  1.00  0.00           C\\nATOM    379   C CYS A  95      -0.350 -18.131 -13.345  1.00  0.00           C\\nATOM    380   O CYS A  95      -0.357 -18.677 -14.449  1.00  0.00           O\\nATOM    381   N CYS A  96      -1.476 -17.906 -12.651  1.00  0.00           N\\nATOM    382 CA  CYS A  96      -2.770 -18.428 -13.075  1.00  0.00           C\\nATOM    383   C CYS A  96      -3.216 -17.784 -14.382  1.00  0.00           C\\nATOM    384   O CYS A  96      -3.667 -18.475 -15.298  1.00  0.00           O\\nATOM    385   N THR A  97      -3.012 -16.453 -14.570  1.00  0.00           N\\nATOM    386 CA  THR A  97      -3.394 -15.726 -15.775  1.00  0.00           C\\nATOM    387   C THR A  97      -2.562 -16.184 -16.970  1.00  0.00           C\\nATOM    388   O THR A  97      -3.096 -16.391 -18.062  1.00  0.00           O\\nATOM    389   N SER A  98      -1.230 -16.382 -16.684  1.00  0.00           N\\nATOM    390 CA  SER A  98      -0.336 -16.877 -17.726  1.00  0.00           C\\nATOM    391   C SER A  98      -0.771 -18.253 -18.220  1.00  0.00           C\\nATOM    392   O SER A  98      -0.842 -18.491 -19.427  1.00  0.00           O\\nATOM    393   N ILE A  99      -1.173 -19.156 -17.331  1.00  0.00           N\\nATOM    394 CA  ILE A  99      -1.585 -20.514 -17.668  1.00  0.00           C\\nATOM    395   C ILE A  99      -2.883 -20.475 -18.470  1.00  0.00           C\\nATOM    396   O ILE A  99      -3.011 -21.157 -19.490  1.00  0.00           O\\nATOM    397   N CYS A 100      -3.881 -19.647 -18.027  1.00  0.00           N\\nATOM    398 CA  CYS A 100      -5.168 -19.523 -18.702  1.00  0.00           C\\nATOM    399   C CYS A 100      -4.991 -18.987 -20.118  1.00  0.00           C\\nATOM    400   O CYS A 100      -5.608 -19.491 -21.058  1.00  0.00           O\\nATOM    401   N SER A 101      -4.058 -18.087 -20.291  1.00  0.00           N\\nATOM    402 CA  SER A 101      -3.775 -17.508 -21.600  1.00  0.00           C\\nATOM    403   C SER A 101      -3.147 -18.536 -22.535  1.00  0.00           C\\nATOM    404   O SER A 101      -3.539 -18.647 -23.698  1.00  0.00           O\\nATOM    405   N LEU A 102      -2.165 -19.253 -22.027  1.00  0.00           N\\nATOM    406 CA  LEU A 102      -1.531 -20.309 -22.808  1.00  0.00           C\\nATOM    407   C LEU A 102      -2.561 -21.333 -23.275  1.00  0.00           C\\nATOM    408   O LEU A 102      -2.563 -21.728 -24.443  1.00  0.00           O\\nATOM    409   N TYR A 103      -3.478 -21.700 -22.375  1.00  0.00           N\\nATOM    410 CA  TYR A 103      -4.517 -22.664 -22.717  1.00  0.00           C\\nATOM    411   C TYR A 103      -5.431 -22.118 -23.807  1.00  0.00           C\\nATOM    412   O TYR A 103      -5.761 -22.825 -24.762  1.00  0.00           O\\nATOM    413   N GLN A 104      -5.821 -20.839 -23.740  1.00  0.00           N\\nATOM    414 CA  GLN A 104      -6.675 -20.218 -24.747  1.00  0.00           C\\nATOM    415   C GLN A 104      -5.989 -20.191 -26.110  1.00  0.00           C\\nATOM    416   O GLN A 104      -6.617 -20.477 -27.132  1.00  0.00           O\\nATOM    417   N LEU A 105      -4.683 -19.911 -26.130  1.00  0.00           N\\nATOM    418 CA  LEU A 105      -3.889 -19.917 -27.354  1.00  0.00           C\\nATOM    419   C LEU A 105      -3.868 -21.306 -27.982  1.00  0.00           C\\nATOM    420   O LEU A 105      -4.061 -21.449 -29.191  1.00  0.00           O\\nATOM    421   N GLU A 106      -3.629 -22.313 -27.147  1.00  0.00           N\\nATOM    422 CA  GLU A 106      -3.606 -23.685 -27.646  1.00  0.00           C\\nATOM    423   C GLU A 106      -4.951 -24.075 -28.252  1.00  0.00           C\\nATOM    424   O GLU A 106      -5.001 -24.695 -29.316  1.00  0.00           O\\nATOM    425   N ASN A 107      -6.100 -23.604 -27.645  1.00  0.00           N\\nATOM    426 CA  ASN A 107      -7.441 -23.945 -28.106  1.00  0.00           C\\nATOM    427   C ASN A 107      -7.780 -23.240 -29.417  1.00  0.00           C\\nATOM    428   O ASN A 107      -8.415 -23.826 -30.295  1.00  0.00           O\\nATOM    429   N TYR A 108      -7.183 -21.998 -29.659  1.00  0.00           N\\nATOM    430 CA  TYR A 108      -7.414 -21.255 -30.893  1.00  0.00           C\\nATOM    431   C TYR A 108      -6.576 -21.820 -32.034  1.00  0.00           C\\nATOM    432   O TYR A 108      -6.977 -21.753 -33.198  1.00  0.00           O\\nATOM    433   N CYS A 109      -5.335 -22.397 -31.767  1.00  0.00           N\\nATOM    434 CA  CYS A 109      -4.467 -23.065 -32.730  1.00  0.00           C\\nATOM    435   C CYS A 109      -5.088 -24.373 -33.208  1.00  0.00           C\\nATOM    436   O CYS A 109      -4.905 -24.766 -34.361  1.00  0.00           O\\nATOM    437   N ASN A 110      -5.952 -25.004 -32.405  1.00  0.00           N\\nATOM    438 CA  ASN A 110      -6.677 -26.192 -32.842  1.00  0.00           C\\nATOM    439   C ASN A 110      -7.916 -25.825 -33.655  1.00  0.00           C\\nATOM    440   O ASN A 110      -8.204 -26.457 -34.673  1.00  0.00           O\\nTER\\nEND\\n\",\"pdb\");\n\tviewer_17613512088053458.setBackgroundColor(\"white\");\n\tviewer_17613512088053458.setStyle({\"cartoon\": {\"color\": \"spectrum\", \"thickness\": 1.2, \"arrows\": true}});\n\tviewer_17613512088053458.addStyle({\"atom\": [\"N\", \"CA\", \"C\", \"O\"]},{\"stick\": {\"radius\": 0.25}});\n\tviewer_17613512088053458.zoomTo();\n\tviewer_17613512088053458.zoom(1.25);\nviewer_17613512088053458.render();\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_17613512088053458\"  style=\"position: relative; width: 900px; height: 600px;\">\n",
       "        <p id=\"3dmolwarning_17613512088053458\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    //this is to ignore the existence of requirejs amd\n",
       "    var savedexports, savedmodule;\n",
       "    if (typeof exports !== 'undefined') savedexports = exports;\n",
       "    else exports = {}\n",
       "    if (typeof module !== 'undefined') savedmodule = module;\n",
       "    else module = {}\n",
       "\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "        exports = savedexports;\n",
       "        module = savedmodule;\n",
       "        resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://cdn.jsdelivr.net/npm/3dmol@2.5.3/build/3Dmol-min.js');\n",
       "}\n",
       "\n",
       "var viewer_17613512088053458 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_17613512088053458\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_17613512088053458 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17613512088053458\"),{backgroundColor:\"white\"});\n",
       "viewer_17613512088053458.zoomTo();\n",
       "\tviewer_17613512088053458.addModel(\"ATOM      1   N MET A   1      -2.291 -14.445 -41.302  1.00  0.00           N\\nATOM      2 CA  MET A   1      -2.968 -13.376 -40.574  1.00  0.00           C\\nATOM      3   C MET A   1      -3.255 -13.795 -39.136  1.00  0.00           C\\nATOM      4   O MET A   1      -3.010 -13.029 -38.202  1.00  0.00           O\\nATOM      5   N ALA A   2      -3.712 -15.067 -38.982  1.00  0.00           N\\nATOM      6 CA  ALA A   2      -4.057 -15.557 -37.651  1.00  0.00           C\\nATOM      7   C ALA A   2      -2.824 -15.626 -36.754  1.00  0.00           C\\nATOM      8   O ALA A   2      -2.886 -15.272 -35.574  1.00  0.00           O\\nATOM      9   N LEU A   3      -1.670 -15.930 -37.250  1.00  0.00           N\\nATOM     10 CA  LEU A   3      -0.412 -16.029 -36.519  1.00  0.00           C\\nATOM     11   C LEU A   3       0.055 -14.655 -36.051  1.00  0.00           C\\nATOM     12   O LEU A   3       0.477 -14.496 -34.903  1.00  0.00           O\\nATOM     13   N TRP A   4      -0.036 -13.714 -36.947  1.00  0.00           N\\nATOM     14 CA  TRP A   4       0.364 -12.346 -36.635  1.00  0.00           C\\nATOM     15   C TRP A   4      -0.495 -11.769 -35.515  1.00  0.00           C\\nATOM     16   O TRP A   4       0.024 -11.147 -34.584  1.00  0.00           O\\nATOM     17   N MET A   5      -1.765 -11.945 -35.634  1.00  0.00           N\\nATOM     18 CA  MET A   5      -2.701 -11.453 -34.627  1.00  0.00           C\\nATOM     19   C MET A   5      -2.411 -12.075 -33.265  1.00  0.00           C\\nATOM     20   O MET A   5      -2.414 -11.380 -32.247  1.00  0.00           O\\nATOM     21   N ARG A   6      -2.081 -13.374 -33.270  1.00  0.00           N\\nATOM     22 CA  ARG A   6      -1.714 -14.095 -32.056  1.00  0.00           C\\nATOM     23   C ARG A   6      -0.435 -13.530 -31.448  1.00  0.00           C\\nATOM     24   O ARG A   6      -0.355 -13.326 -30.235  1.00  0.00           O\\nATOM     25   N LEU A   7       0.480 -13.328 -32.248  1.00  0.00           N\\nATOM     26 CA  LEU A   7       1.756 -12.767 -31.818  1.00  0.00           C\\nATOM     27   C LEU A   7       1.565 -11.376 -31.222  1.00  0.00           C\\nATOM     28   O LEU A   7       2.123 -11.065 -30.167  1.00  0.00           O\\nATOM     29   N LEU A   8       0.760 -10.583 -31.831  1.00  0.00           N\\nATOM     30 CA  LEU A   8       0.481  -9.226 -31.372  1.00  0.00           C\\nATOM     31   C LEU A   8      -0.221  -9.244 -30.019  1.00  0.00           C\\nATOM     32   O LEU A   8       0.138  -8.483 -29.118  1.00  0.00           O\\nATOM     33   N PRO A   9      -1.223 -10.139 -29.816  1.00  0.00           N\\nATOM     34 CA  PRO A   9      -1.909 -10.285 -28.530  1.00  0.00           C\\nATOM     35   C PRO A   9      -0.971 -10.727 -27.409  1.00  0.00           C\\nATOM     36   O PRO A   9      -1.044 -10.202 -26.295  1.00  0.00           O\\nATOM     37   N LEU A  10      -0.100 -11.685 -27.713  1.00  0.00           N\\nATOM     38 CA  LEU A  10       0.893 -12.153 -26.752  1.00  0.00           C\\nATOM     39   C LEU A  10       1.821 -11.018 -26.334  1.00  0.00           C\\nATOM     40   O LEU A  10       2.089 -10.836 -25.144  1.00  0.00           O\\nATOM     41   N LEU A  11       2.251 -10.277 -27.318  1.00  0.00           N\\nATOM     42 CA  LEU A  11       3.143  -9.152 -27.055  1.00  0.00           C\\nATOM     43   C LEU A  11       2.455  -8.105 -26.186  1.00  0.00           C\\nATOM     44   O LEU A  11       3.039  -7.618 -25.215  1.00  0.00           O\\nATOM     45   N ALA A  12       1.297  -7.758 -26.530  1.00  0.00           N\\nATOM     46 CA  ALA A  12       0.523  -6.775 -25.776  1.00  0.00           C\\nATOM     47   C ALA A  12       0.300  -7.237 -24.339  1.00  0.00           C\\nATOM     48   O ALA A  12       0.470  -6.458 -23.397  1.00  0.00           O\\nATOM     49   N LEU A  13      -0.003  -8.503 -24.126  1.00  0.00           N\\nATOM     50 CA  LEU A  13      -0.230  -9.077 -22.804  1.00  0.00           C\\nATOM     51   C LEU A  13       1.054  -9.071 -21.981  1.00  0.00           C\\nATOM     52   O LEU A  13       1.036  -8.722 -20.798  1.00  0.00           O\\nATOM     53   N LEU A  14       2.071  -9.472 -22.642  1.00  0.00           N\\nATOM     54 CA  LEU A  14       3.378  -9.443 -21.995  1.00  0.00           C\\nATOM     55   C LEU A  14       3.734  -8.028 -21.551  1.00  0.00           C\\nATOM     56   O LEU A  14       4.184  -7.822 -20.422  1.00  0.00           O\\nATOM     57   N ALA A  15       3.527  -7.083 -22.407  1.00  0.00           N\\nATOM     58 CA  ALA A  15       3.797  -5.681 -22.098  1.00  0.00           C\\nATOM     59   C ALA A  15       2.952  -5.208 -20.919  1.00  0.00           C\\nATOM     60   O ALA A  15       3.462  -4.559 -20.002  1.00  0.00           O\\nATOM     61   N LEU A  16       1.648  -5.528 -20.870  1.00  0.00           N\\nATOM     62 CA  LEU A  16       0.747  -5.163 -19.783  1.00  0.00           C\\nATOM     63   C LEU A  16       1.215  -5.766 -18.463  1.00  0.00           C\\nATOM     64   O LEU A  16       1.229  -5.086 -17.434  1.00  0.00           O\\nATOM     65   N TRP A  17       1.663  -6.991 -18.518  1.00  0.00           N\\nATOM     66 CA  TRP A  17       2.203  -7.688 -17.355  1.00  0.00           C\\nATOM     67   C TRP A  17       3.441  -6.977 -16.820  1.00  0.00           C\\nATOM     68   O TRP A  17       3.562  -6.752 -15.613  1.00  0.00           O\\nATOM     69   N GLY A  18       4.367  -6.726 -17.673  1.00  0.00           N\\nATOM     70 CA  GLY A  18       5.560  -5.988 -17.292  1.00  0.00           C\\nATOM     71   C GLY A  18       5.253  -4.652 -16.641  1.00  0.00           C\\nATOM     72   O GLY A  18       5.836  -4.310 -15.610  1.00  0.00           O\\nATOM     73   N PRO A  19       4.364  -3.982 -17.266  1.00  0.00           N\\nATOM     74 CA  PRO A  19       3.963  -2.687 -16.711  1.00  0.00           C\\nATOM     75   C PRO A  19       3.336  -2.808 -15.324  1.00  0.00           C\\nATOM     76   O PRO A  19       3.680  -2.045 -14.418  1.00  0.00           O\\nATOM     77   N ASP A  20       2.440  -3.795 -15.088  1.00  0.00           N\\nATOM     78 CA  ASP A  20       1.799  -4.027 -13.797  1.00  0.00           C\\nATOM     79   C ASP A  20       2.828  -4.399 -12.732  1.00  0.00           C\\nATOM     80   O ASP A  20       2.778  -3.892 -11.609  1.00  0.00           O\\nATOM     81   N PRO A  21       3.795  -5.267 -13.148  1.00  0.00           N\\nATOM     82 CA  PRO A  21       4.894  -5.611 -12.244  1.00  0.00           C\\nATOM     83   C PRO A  21       5.720  -4.396 -11.829  1.00  0.00           C\\nATOM     84   O PRO A  21       6.021  -4.224 -10.645  1.00  0.00           O\\nATOM     85   N ALA A  22       6.045  -3.550 -12.828  1.00  0.00           N\\nATOM     86 CA  ALA A  22       6.843  -2.354 -12.566  1.00  0.00           C\\nATOM     87   C ALA A  22       6.104  -1.395 -11.637  1.00  0.00           C\\nATOM     88   O ALA A  22       6.674  -0.905 -10.659  1.00  0.00           O\\nATOM     89   N ALA A  23       4.887  -1.130 -11.959  1.00  0.00           N\\nATOM     90 CA  ALA A  23       4.075  -0.226 -11.149  1.00  0.00           C\\nATOM     91   C ALA A  23       3.962  -0.729  -9.713  1.00  0.00           C\\nATOM     92   O ALA A  23       4.121   0.043  -8.764  1.00  0.00           O\\nATOM     93   N ALA A  24       3.762  -1.991  -9.491  1.00  0.00           N\\nATOM     94 CA  ALA A  24       3.658  -2.611  -8.173  1.00  0.00           C\\nATOM     95   C ALA A  24       4.970  -2.488  -7.404  1.00  0.00           C\\nATOM     96   O ALA A  24       4.973  -2.140  -6.221  1.00  0.00           O\\nATOM     97   N PHE A  25       6.027  -2.842  -8.092  1.00  0.00           N\\nATOM     98 CA  PHE A  25       7.356  -2.692  -7.510  1.00  0.00           C\\nATOM     99   C PHE A  25       7.579  -1.263  -7.031  1.00  0.00           C\\nATOM    100   O PHE A  25       8.021  -1.043  -5.901  1.00  0.00           O\\nATOM    101   N VAL A  26       7.221  -0.278  -7.839  1.00  0.00           N\\nATOM    102 CA  VAL A  26       7.403   1.134  -7.518  1.00  0.00           C\\nATOM    103   C VAL A  26       6.556   1.501  -6.301  1.00  0.00           C\\nATOM    104   O VAL A  26       7.045   2.137  -5.365  1.00  0.00           O\\nATOM    105   N ASN A  27       5.273   1.140  -6.310  1.00  0.00           N\\nATOM    106 CA  ASN A  27       4.374   1.420  -5.195  1.00  0.00           C\\nATOM    107   C ASN A  27       4.889   0.812  -3.894  1.00  0.00           C\\nATOM    108   O ASN A  27       4.868   1.463  -2.848  1.00  0.00           O\\nATOM    109   N GLN A  28       5.411  -0.399  -3.977  1.00  0.00           N\\nATOM    110 CA  GLN A  28       5.980  -1.105  -2.834  1.00  0.00           C\\nATOM    111   C GLN A  28       7.199  -0.370  -2.284  1.00  0.00           C\\nATOM    112   O GLN A  28       7.334  -0.203  -1.070  1.00  0.00           O\\nATOM    113   N HIS A  29       8.128  -0.030  -3.200  1.00  0.00           N\\nATOM    114 CA  HIS A  29       9.320   0.725  -2.831  1.00  0.00           C\\nATOM    115   C HIS A  29       8.951   2.026  -2.127  1.00  0.00           C\\nATOM    116   O HIS A  29       9.526   2.360  -1.088  1.00  0.00           O\\nATOM    117   N LEU A  30       7.976   2.763  -2.637  1.00  0.00           N\\nATOM    118 CA  LEU A  30       7.540   4.036  -2.074  1.00  0.00           C\\nATOM    119   C LEU A  30       6.967   3.843  -0.674  1.00  0.00           C\\nATOM    120   O LEU A  30       7.316   4.579   0.252  1.00  0.00           O\\nATOM    121   N CYS A  31       6.071   2.774  -0.424  1.00  0.00           N\\nATOM    122 CA  CYS A  31       5.470   2.468   0.869  1.00  0.00           C\\nATOM    123   C CYS A  31       6.538   2.111   1.896  1.00  0.00           C\\nATOM    124   O CYS A  31       6.503   2.595   3.028  1.00  0.00           O\\nATOM    125   N GLY A  32       7.541   1.231   1.426  1.00  0.00           N\\nATOM    126 CA  GLY A  32       8.670   0.909   2.284  1.00  0.00           C\\nATOM    127   C GLY A  32       9.466   2.129   2.706  1.00  0.00           C\\nATOM    128   O GLY A  32       9.792   2.287   3.884  1.00  0.00           O\\nATOM    129   N SER A  33       9.755   3.026   1.763  1.00  0.00           N\\nATOM    130 CA  SER A  33      10.520   4.239   2.032  1.00  0.00           C\\nATOM    131   C SER A  33       9.796   5.141   3.026  1.00  0.00           C\\nATOM    132   O SER A  33      10.400   5.632   3.981  1.00  0.00           O\\nATOM    133   N HIS A  34       8.517   5.416   2.789  1.00  0.00           N\\nATOM    134 CA  HIS A  34       7.711   6.255   3.668  1.00  0.00           C\\nATOM    135   C HIS A  34       7.686   5.702   5.089  1.00  0.00           C\\nATOM    136   O HIS A  34       7.826   6.455   6.055  1.00  0.00           O\\nATOM    137   N LEU A  35       7.637   4.366   5.194  1.00  0.00           N\\nATOM    138 CA  LEU A  35       7.601   3.720   6.502  1.00  0.00           C\\nATOM    139   C LEU A  35       8.935   3.877   7.223  1.00  0.00           C\\nATOM    140   O LEU A  35       8.968   4.202   8.413  1.00  0.00           O\\nATOM    141   N VAL A  36       9.998   3.641   6.490  1.00  0.00           N\\nATOM    142 CA  VAL A  36      11.341   3.820   7.031  1.00  0.00           C\\nATOM    143   C VAL A  36      11.495   5.239   7.573  1.00  0.00           C\\nATOM    144   O VAL A  36      11.984   5.435   8.688  1.00  0.00           O\\nATOM    145   N GLU A  37      11.034   6.194   6.789  1.00  0.00           N\\nATOM    146 CA  GLU A  37      11.116   7.595   7.191  1.00  0.00           C\\nATOM    147   C GLU A  37      10.302   7.854   8.456  1.00  0.00           C\\nATOM    148   O GLU A  37      10.783   8.502   9.388  1.00  0.00           O\\nATOM    149   N ALA A  38       9.074   7.347   8.500  1.00  0.00           N\\nATOM    150 CA  ALA A  38       8.211   7.508   9.667  1.00  0.00           C\\nATOM    151   C ALA A  38       8.850   6.897  10.911  1.00  0.00           C\\nATOM    152   O ALA A  38       8.836   7.503  11.985  1.00  0.00           O\\nATOM    153   N LEU A  39       9.461   5.702  10.755  1.00  0.00           N\\nATOM    154 CA  LEU A  39      10.163   5.023  11.839  1.00  0.00           C\\nATOM    155   C LEU A  39      11.324   5.871  12.349  1.00  0.00           C\\nATOM    156   O LEU A  39      11.492   6.037  13.559  1.00  0.00           O\\nATOM    157   N TYR A  40      12.132   6.392  11.416  1.00  0.00           N\\nATOM    158 CA  TYR A  40      13.263   7.239  11.777  1.00  0.00           C\\nATOM    159   C TYR A  40      12.801   8.465  12.555  1.00  0.00           C\\nATOM    160   O TYR A  40      13.380   8.806  13.589  1.00  0.00           O\\nATOM    161   N LEU A  41      11.712   9.114  12.128  1.00  0.00           N\\nATOM    162 CA  LEU A  41      11.185  10.310  12.776  1.00  0.00           C\\nATOM    163   C LEU A  41      10.718   9.998  14.194  1.00  0.00           C\\nATOM    164   O LEU A  41      11.036  10.732  15.132  1.00  0.00           O\\nATOM    165   N VAL A  42      10.022   8.859  14.375  1.00  0.00           N\\nATOM    166 CA  VAL A  42       9.500   8.463  15.678  1.00  0.00           C\\nATOM    167   C VAL A  42      10.656   8.173  16.632  1.00  0.00           C\\nATOM    168   O VAL A  42      10.626   8.578  17.796  1.00  0.00           O\\nATOM    169   N CYS A  43      11.669   7.462  16.099  1.00  0.00           N\\nATOM    170 CA  CYS A  43      12.864   7.175  16.885  1.00  0.00           C\\nATOM    171   C CYS A  43      13.544   8.463  17.333  1.00  0.00           C\\nATOM    172   O CYS A  43      13.907   8.603  18.503  1.00  0.00           O\\nATOM    173   N GLY A  44      13.711   9.430  16.400  1.00  0.00           N\\nATOM    174 CA  GLY A  44      14.318  10.706  16.744  1.00  0.00           C\\nATOM    175   C GLY A  44      13.535  11.477  17.790  1.00  0.00           C\\nATOM    176   O GLY A  44      14.113  11.996  18.748  1.00  0.00           O\\nATOM    177   N GLU A  45      12.221  11.581  17.581  1.00  0.00           N\\nATOM    178 CA  GLU A  45      11.365  12.295  18.524  1.00  0.00           C\\nATOM    179   C GLU A  45      11.451  11.683  19.919  1.00  0.00           C\\nATOM    180   O GLU A  45      11.524  12.405  20.916  1.00  0.00           O\\nATOM    181   N ARG A  46      11.588  10.390  19.961  1.00  0.00           N\\nATOM    182 CA  ARG A  46      11.727   9.676  21.226  1.00  0.00           C\\nATOM    183   C ARG A  46      13.059  10.000  21.895  1.00  0.00           C\\nATOM    184   O ARG A  46      13.113  10.223  23.106  1.00  0.00           O\\nATOM    185   N GLY A  47      14.107   9.893  21.074  1.00  0.00           N\\nATOM    186 CA  GLY A  47      15.407  10.282  21.598  1.00  0.00           C\\nATOM    187   C GLY A  47      15.421  11.683  22.179  1.00  0.00           C\\nATOM    188   O GLY A  47      15.950  11.903  23.271  1.00  0.00           O\\nATOM    189   N PHE A  48      14.808  12.688  21.455  1.00  0.00           N\\nATOM    190 CA  PHE A  48      14.762  14.067  21.925  1.00  0.00           C\\nATOM    191   C PHE A  48      14.000  14.166  23.241  1.00  0.00           C\\nATOM    192   O PHE A  48      14.451  14.829  24.177  1.00  0.00           O\\nATOM    193   N PHE A  49      12.816  13.449  23.425  1.00  0.00           N\\nATOM    194 CA  PHE A  49      12.005  13.460  24.637  1.00  0.00           C\\nATOM    195   C PHE A  49      12.784  12.886  25.815  1.00  0.00           C\\nATOM    196   O PHE A  49      12.740  13.431  26.920  1.00  0.00           O\\nATOM    197   N TYR A  50      13.522  11.760  25.492  1.00  0.00           N\\nATOM    198 CA  TYR A  50      14.402  11.156  26.487  1.00  0.00           C\\nATOM    199   C TYR A  50      15.404  12.174  27.018  1.00  0.00           C\\nATOM    200   O TYR A  50      15.616  12.271  28.229  1.00  0.00           O\\nATOM    201   N THR A  51      16.073  12.781  26.080  1.00  0.00           N\\nATOM    202 CA  THR A  51      17.091  13.740  26.493  1.00  0.00           C\\nATOM    203   C THR A  51      16.486  14.822  27.384  1.00  0.00           C\\nATOM    204   O THR A  51      17.083  15.211  28.389  1.00  0.00           O\\nATOM    205   N PRO A  52      15.184  15.129  27.036  1.00  0.00           N\\nATOM    206 CA  PRO A  52      14.560  16.214  27.797  1.00  0.00           C\\nATOM    207   C PRO A  52      14.018  15.751  29.148  1.00  0.00           C\\nATOM    208   O PRO A  52      14.094  16.489  30.133  1.00  0.00           O\\nATOM    209   N LYS A  53      13.486  14.345  29.353  1.00  0.00           N\\nATOM    210 CA  LYS A  53      12.801  13.884  30.557  1.00  0.00           C\\nATOM    211   C LYS A  53      13.772  13.200  31.515  1.00  0.00           C\\nATOM    212   O LYS A  53      13.501  13.095  32.713  1.00  0.00           O\\nATOM    213   N THR A  54      14.952  12.698  31.095  1.00  0.00           N\\nATOM    214 CA  THR A  54      16.048  12.218  31.930  1.00  0.00           C\\nATOM    215   C THR A  54      16.214  13.102  33.163  1.00  0.00           C\\nATOM    216   O THR A  54      16.613  12.623  34.227  1.00  0.00           O\\nATOM    217   N ARG A  55      15.375  14.042  33.280  1.00  0.00           N\\nATOM    218 CA  ARG A  55      15.565  14.890  34.452  1.00  0.00           C\\nATOM    219   C ARG A  55      14.581  14.525  35.557  1.00  0.00           C\\nATOM    220   O ARG A  55      14.903  14.632  36.743  1.00  0.00           O\\nATOM    221   N ARG A  56      13.737  13.378  35.456  1.00  0.00           N\\nATOM    222 CA  ARG A  56      12.809  13.162  36.561  1.00  0.00           C\\nATOM    223   C ARG A  56      12.154  11.788  36.466  1.00  0.00           C\\nATOM    224   O ARG A  56      11.640  11.411  35.411  1.00  0.00           O\\nATOM    225   N GLU A  57      12.686  10.766  36.986  1.00  0.00           N\\nATOM    226 CA  GLU A  57      12.189   9.461  37.411  1.00  0.00           C\\nATOM    227   C GLU A  57      12.541   8.379  36.394  1.00  0.00           C\\nATOM    228   O GLU A  57      11.980   8.346  35.296  1.00  0.00           O\\nATOM    229   N ALA A  58      13.771   8.025  36.192  1.00  0.00           N\\nATOM    230 CA  ALA A  58      14.596   7.225  35.291  1.00  0.00           C\\nATOM    231   C ALA A  58      13.975   5.851  35.055  1.00  0.00           C\\nATOM    232   O ALA A  58      13.959   5.355  33.926  1.00  0.00           O\\nATOM    233   N GLU A  59      13.284   5.324  36.030  1.00  0.00           N\\nATOM    234 CA  GLU A  59      12.908   3.917  35.930  1.00  0.00           C\\nATOM    235   C GLU A  59      11.584   3.750  35.190  1.00  0.00           C\\nATOM    236   O GLU A  59      11.475   2.925  34.281  1.00  0.00           O\\nATOM    237   N ASP A  60      10.442   4.398  35.561  1.00  0.00           N\\nATOM    238 CA  ASP A  60       9.159   4.333  34.869  1.00  0.00           C\\nATOM    239   C ASP A  60       9.310   4.702  33.395  1.00  0.00           C\\nATOM    240   O ASP A  60       8.690   4.083  32.527  1.00  0.00           O\\nATOM    241   N LEU A  61      10.212   5.430  33.206  1.00  0.00           N\\nATOM    242 CA  LEU A  61      10.460   5.928  31.857  1.00  0.00           C\\nATOM    243   C LEU A  61      11.108   4.852  30.991  1.00  0.00           C\\nATOM    244   O LEU A  61      10.736   4.678  29.828  1.00  0.00           O\\nATOM    245   N GLN A  62      12.024   4.113  31.525  1.00  0.00           N\\nATOM    246 CA  GLN A  62      12.672   2.986  30.862  1.00  0.00           C\\nATOM    247   C GLN A  62      11.648   1.946  30.418  1.00  0.00           C\\nATOM    248   O GLN A  62      11.694   1.466  29.284  1.00  0.00           O\\nATOM    249   N VAL A  63      10.735   1.656  31.327  1.00  0.00           N\\nATOM    250 CA  VAL A  63       9.694   0.675  31.037  1.00  0.00           C\\nATOM    251   C VAL A  63       8.821   1.172  29.887  1.00  0.00           C\\nATOM    252   O VAL A  63       8.515   0.417  28.961  1.00  0.00           O\\nATOM    253   N GLY A  64       8.385   2.421  29.917  1.00  0.00           N\\nATOM    254 CA  GLY A  64       7.606   2.997  28.833  1.00  0.00           C\\nATOM    255   C GLY A  64       8.330   2.975  27.500  1.00  0.00           C\\nATOM    256   O GLY A  64       7.728   2.679  26.465  1.00  0.00           O\\nATOM    257   N GLN A  65       9.598   3.204  27.499  1.00  0.00           N\\nATOM    258 CA  GLN A  65      10.452   3.144  26.318  1.00  0.00           C\\nATOM    259   C GLN A  65      10.489   1.734  25.736  1.00  0.00           C\\nATOM    260   O GLN A  65      10.363   1.553  24.523  1.00  0.00           O\\nATOM    261   N VAL A  66      10.730   0.853  26.582  1.00  0.00           N\\nATOM    262 CA  VAL A  66      10.774  -0.549  26.181  1.00  0.00           C\\nATOM    263   C VAL A  66       9.450  -0.942  25.530  1.00  0.00           C\\nATOM    264   O VAL A  66       9.435  -1.577  24.473  1.00  0.00           O\\nATOM    265   N GLU A  67       8.387  -0.510  26.073  1.00  0.00           N\\nATOM    266 CA  GLU A  67       7.061  -0.854  25.568  1.00  0.00           C\\nATOM    267   C GLU A  67       6.807  -0.217  24.205  1.00  0.00           C\\nATOM    268   O GLU A  67       6.307  -0.875  23.290  1.00  0.00           O\\nATOM    269   N LEU A  68       7.155   1.030  24.073  1.00  0.00           N\\nATOM    270 CA  LEU A  68       6.966   1.752  22.819  1.00  0.00           C\\nATOM    271   C LEU A  68       7.820   1.147  21.710  1.00  0.00           C\\nATOM    272   O LEU A  68       7.351   0.978  20.582  1.00  0.00           O\\nATOM    273   N GLY A  69       9.144   0.915  22.015  1.00  0.00           N\\nATOM    274 CA  GLY A  69      10.011   0.200  21.092  1.00  0.00           C\\nATOM    275   C GLY A  69       9.447  -1.140  20.660  1.00  0.00           C\\nATOM    276   O GLY A  69       9.462  -1.473  19.473  1.00  0.00           O\\nATOM    277   N GLY A  70       8.946  -1.902  21.656  1.00  0.00           N\\nATOM    278 CA  GLY A  70       8.326  -3.180  21.343  1.00  0.00           C\\nATOM    279   C GLY A  70       7.128  -3.052  20.422  1.00  0.00           C\\nATOM    280   O GLY A  70       6.999  -3.805  19.454  1.00  0.00           O\\nATOM    281   N GLY A  71       6.206  -2.243  20.712  1.00  0.00           N\\nATOM    282 CA  GLY A  71       5.058  -1.993  19.855  1.00  0.00           C\\nATOM    283   C GLY A  71       5.442  -1.629  18.433  1.00  0.00           C\\nATOM    284   O GLY A  71       4.850  -2.133  17.477  1.00  0.00           O\\nATOM    285   N PRO A  72       6.390  -0.768  18.246  1.00  0.00           N\\nATOM    286 CA  PRO A  72       6.884  -0.400  16.917  1.00  0.00           C\\nATOM    287   C PRO A  72       7.444  -1.593  16.145  1.00  0.00           C\\nATOM    288   O PRO A  72       7.153  -1.756  14.957  1.00  0.00           O\\nATOM    289   N GLY A  73       8.363  -2.260  16.892  1.00  0.00           N\\nATOM    290 CA  GLY A  73       8.868  -3.481  16.286  1.00  0.00           C\\nATOM    291   C GLY A  73       7.767  -4.410  15.810  1.00  0.00           C\\nATOM    292   O GLY A  73       7.819  -4.920  14.689  1.00  0.00           O\\nATOM    293   N ALA A  74       6.816  -4.578  16.633  1.00  0.00           N\\nATOM    294 CA  ALA A  74       5.692  -5.447  16.292  1.00  0.00           C\\nATOM    295   C ALA A  74       4.946  -4.922  15.068  1.00  0.00           C\\nATOM    296   O ALA A  74       4.627  -5.685  14.154  1.00  0.00           O\\nATOM    297   N GLY A  75       4.575  -3.587  14.998  1.00  0.00           N\\nATOM    298 CA  GLY A  75       3.943  -2.975  13.840  1.00  0.00           C\\nATOM    299   C GLY A  75       4.737  -3.156  12.560  1.00  0.00           C\\nATOM    300   O GLY A  75       4.173  -3.482  11.514  1.00  0.00           O\\nATOM    301   N SER A  76       6.093  -3.023  12.656  1.00  0.00           N\\nATOM    302 CA  SER A  76       6.984  -3.243  11.521  1.00  0.00           C\\nATOM    303   C SER A  76       6.896  -4.681  11.019  1.00  0.00           C\\nATOM    304   O SER A  76       6.797  -4.918   9.814  1.00  0.00           O\\nATOM    305   N LEU A  77       6.971  -5.592  11.997  1.00  0.00           N\\nATOM    306 CA  LEU A  77       6.894  -7.008  11.656  1.00  0.00           C\\nATOM    307   C LEU A  77       5.576  -7.329  10.959  1.00  0.00           C\\nATOM    308   O LEU A  77       5.562  -8.018   9.936  1.00  0.00           O\\nATOM    309   N GLN A  78       4.433  -6.891  11.513  1.00  0.00           N\\nATOM    310 CA  GLN A  78       3.108  -7.127  10.948  1.00  0.00           C\\nATOM    311   C GLN A  78       3.011  -6.580   9.527  1.00  0.00           C\\nATOM    312   O GLN A  78       2.496  -7.253   8.631  1.00  0.00           O\\nATOM    313   N PRO A  79       3.546  -5.435   9.267  1.00  0.00           N\\nATOM    314 CA  PRO A  79       3.511  -4.840   7.929  1.00  0.00           C\\nATOM    315   C PRO A  79       4.351  -5.616   6.916  1.00  0.00           C\\nATOM    316   O PRO A  79       3.908  -5.843   5.787  1.00  0.00           O\\nATOM    317   N LEU A  80       5.572  -5.953   7.372  1.00  0.00           N\\nATOM    318 CA  LEU A  80       6.422  -6.803   6.545  1.00  0.00           C\\nATOM    319   C LEU A  80       5.689  -8.079   6.145  1.00  0.00           C\\nATOM    320   O LEU A  80       5.728  -8.484   4.981  1.00  0.00           O\\nATOM    321   N ALA A  81       4.986  -8.697   7.140  1.00  0.00           N\\nATOM    322 CA  ALA A  81       4.232  -9.922   6.888  1.00  0.00           C\\nATOM    323   C ALA A  81       3.116  -9.681   5.875  1.00  0.00           C\\nATOM    324   O ALA A  81       2.940 -10.463   4.938  1.00  0.00           O\\nATOM    325   N LEU A  82       2.387  -8.605   6.024  1.00  0.00           N\\nATOM    326 CA  LEU A  82       1.311  -8.232   5.113  1.00  0.00           C\\nATOM    327   C LEU A  82       1.843  -8.028   3.698  1.00  0.00           C\\nATOM    328   O LEU A  82       1.249  -8.512   2.731  1.00  0.00           O\\nATOM    329   N GLU A  83       3.004  -7.406   3.571  1.00  0.00           N\\nATOM    330 CA  GLU A  83       3.668  -7.167   2.293  1.00  0.00           C\\nATOM    331   C GLU A  83       4.048  -8.480   1.615  1.00  0.00           C\\nATOM    332   O GLU A  83       3.810  -8.659   0.419  1.00  0.00           O\\nATOM    333   N GLY A  84       4.736  -9.356   2.399  1.00  0.00           N\\nATOM    334 CA  GLY A  84       5.068 -10.676   1.885  1.00  0.00           C\\nATOM    335   C GLY A  84       3.861 -11.436   1.370  1.00  0.00           C\\nATOM    336   O GLY A  84       3.905 -12.019   0.284  1.00  0.00           O\\nATOM    337   N SER A  85       2.763 -11.409   2.108  1.00  0.00           N\\nATOM    338 CA  SER A  85       1.534 -12.103   1.736  1.00  0.00           C\\nATOM    339   C SER A  85       0.948 -11.537   0.447  1.00  0.00           C\\nATOM    340   O SER A  85       0.566 -12.290  -0.450  1.00  0.00           O\\nATOM    341   N LEU A  86       0.928 -10.184   0.255  1.00  0.00           N\\nATOM    342 CA  LEU A  86       0.423  -9.512  -0.937  1.00  0.00           C\\nATOM    343   C LEU A  86       1.251  -9.882  -2.163  1.00  0.00           C\\nATOM    344   O LEU A  86       0.699 -10.163  -3.229  1.00  0.00           O\\nATOM    345   N GLN A  87       2.645  -9.925  -2.028  1.00  0.00           N\\nATOM    346 CA  GLN A  87       3.545 -10.342  -3.098  1.00  0.00           C\\nATOM    347   C GLN A  87       3.241 -11.768  -3.549  1.00  0.00           C\\nATOM    348   O GLN A  87       3.156 -12.040  -4.748  1.00  0.00           O\\nATOM    349   N LYS A  88       3.087 -12.683  -2.535  1.00  0.00           N\\nATOM    350 CA  LYS A  88       2.812 -14.084  -2.842  1.00  0.00           C\\nATOM    351   C LYS A  88       1.505 -14.231  -3.616  1.00  0.00           C\\nATOM    352   O LYS A  88       1.450 -14.942  -4.622  1.00  0.00           O\\nATOM    353   N ARG A  89       0.396 -13.590  -3.183  1.00  0.00           N\\nATOM    354 CA  ARG A  89      -0.914 -13.645  -3.824  1.00  0.00           C\\nATOM    355   C ARG A  89      -0.843 -13.139  -5.261  1.00  0.00           C\\nATOM    356   O ARG A  89      -1.398 -13.757  -6.171  1.00  0.00           O\\nATOM    357   N GLY A  90      -0.100 -12.084  -5.475  1.00  0.00           N\\nATOM    358 CA  GLY A  90       0.060 -11.542  -6.814  1.00  0.00           C\\nATOM    359   C GLY A  90       0.796 -12.479  -7.753  1.00  0.00           C\\nATOM    360   O GLY A  90       0.383 -12.669  -8.899  1.00  0.00           O\\nATOM    361   N ILE A  91       1.941 -13.038  -7.220  1.00  0.00           N\\nATOM    362 CA  ILE A  91       2.699 -14.013  -7.996  1.00  0.00           C\\nATOM    363   C ILE A  91       1.783 -15.156  -8.425  1.00  0.00           C\\nATOM    364   O ILE A  91       1.817 -15.588  -9.579  1.00  0.00           O\\nATOM    365   N VAL A  92       0.903 -15.641  -7.527  1.00  0.00           N\\nATOM    366 CA  VAL A  92      -0.018 -16.741  -7.798  1.00  0.00           C\\nATOM    367   C VAL A  92      -1.019 -16.324  -8.872  1.00  0.00           C\\nATOM    368   O VAL A  92      -1.263 -17.068  -9.825  1.00  0.00           O\\nATOM    369   N GLU A  93      -1.605 -15.091  -8.735  1.00  0.00           N\\nATOM    370 CA  GLU A  93      -2.576 -14.576  -9.695  1.00  0.00           C\\nATOM    371   C GLU A  93      -1.961 -14.442 -11.085  1.00  0.00           C\\nATOM    372   O GLU A  93      -2.584 -14.808 -12.084  1.00  0.00           O\\nATOM    373   N GLN A  94      -0.683 -14.027 -11.125  1.00  0.00           N\\nATOM    374 CA  GLN A  94       0.067 -13.903 -12.370  1.00  0.00           C\\nATOM    375   C GLN A  94       0.263 -15.263 -13.033  1.00  0.00           C\\nATOM    376   O GLN A  94       0.063 -15.406 -14.241  1.00  0.00           O\\nATOM    377   N CYS A  95       0.732 -16.205 -12.268  1.00  0.00           N\\nATOM    378 CA  CYS A  95       0.941 -17.557 -12.774  1.00  0.00           C\\nATOM    379   C CYS A  95      -0.350 -18.131 -13.345  1.00  0.00           C\\nATOM    380   O CYS A  95      -0.357 -18.677 -14.449  1.00  0.00           O\\nATOM    381   N CYS A  96      -1.476 -17.906 -12.651  1.00  0.00           N\\nATOM    382 CA  CYS A  96      -2.770 -18.428 -13.075  1.00  0.00           C\\nATOM    383   C CYS A  96      -3.216 -17.784 -14.382  1.00  0.00           C\\nATOM    384   O CYS A  96      -3.667 -18.475 -15.298  1.00  0.00           O\\nATOM    385   N THR A  97      -3.012 -16.453 -14.570  1.00  0.00           N\\nATOM    386 CA  THR A  97      -3.394 -15.726 -15.775  1.00  0.00           C\\nATOM    387   C THR A  97      -2.562 -16.184 -16.970  1.00  0.00           C\\nATOM    388   O THR A  97      -3.096 -16.391 -18.062  1.00  0.00           O\\nATOM    389   N SER A  98      -1.230 -16.382 -16.684  1.00  0.00           N\\nATOM    390 CA  SER A  98      -0.336 -16.877 -17.726  1.00  0.00           C\\nATOM    391   C SER A  98      -0.771 -18.253 -18.220  1.00  0.00           C\\nATOM    392   O SER A  98      -0.842 -18.491 -19.427  1.00  0.00           O\\nATOM    393   N ILE A  99      -1.173 -19.156 -17.331  1.00  0.00           N\\nATOM    394 CA  ILE A  99      -1.585 -20.514 -17.668  1.00  0.00           C\\nATOM    395   C ILE A  99      -2.883 -20.475 -18.470  1.00  0.00           C\\nATOM    396   O ILE A  99      -3.011 -21.157 -19.490  1.00  0.00           O\\nATOM    397   N CYS A 100      -3.881 -19.647 -18.027  1.00  0.00           N\\nATOM    398 CA  CYS A 100      -5.168 -19.523 -18.702  1.00  0.00           C\\nATOM    399   C CYS A 100      -4.991 -18.987 -20.118  1.00  0.00           C\\nATOM    400   O CYS A 100      -5.608 -19.491 -21.058  1.00  0.00           O\\nATOM    401   N SER A 101      -4.058 -18.087 -20.291  1.00  0.00           N\\nATOM    402 CA  SER A 101      -3.775 -17.508 -21.600  1.00  0.00           C\\nATOM    403   C SER A 101      -3.147 -18.536 -22.535  1.00  0.00           C\\nATOM    404   O SER A 101      -3.539 -18.647 -23.698  1.00  0.00           O\\nATOM    405   N LEU A 102      -2.165 -19.253 -22.027  1.00  0.00           N\\nATOM    406 CA  LEU A 102      -1.531 -20.309 -22.808  1.00  0.00           C\\nATOM    407   C LEU A 102      -2.561 -21.333 -23.275  1.00  0.00           C\\nATOM    408   O LEU A 102      -2.563 -21.728 -24.443  1.00  0.00           O\\nATOM    409   N TYR A 103      -3.478 -21.700 -22.375  1.00  0.00           N\\nATOM    410 CA  TYR A 103      -4.517 -22.664 -22.717  1.00  0.00           C\\nATOM    411   C TYR A 103      -5.431 -22.118 -23.807  1.00  0.00           C\\nATOM    412   O TYR A 103      -5.761 -22.825 -24.762  1.00  0.00           O\\nATOM    413   N GLN A 104      -5.821 -20.839 -23.740  1.00  0.00           N\\nATOM    414 CA  GLN A 104      -6.675 -20.218 -24.747  1.00  0.00           C\\nATOM    415   C GLN A 104      -5.989 -20.191 -26.110  1.00  0.00           C\\nATOM    416   O GLN A 104      -6.617 -20.477 -27.132  1.00  0.00           O\\nATOM    417   N LEU A 105      -4.683 -19.911 -26.130  1.00  0.00           N\\nATOM    418 CA  LEU A 105      -3.889 -19.917 -27.354  1.00  0.00           C\\nATOM    419   C LEU A 105      -3.868 -21.306 -27.982  1.00  0.00           C\\nATOM    420   O LEU A 105      -4.061 -21.449 -29.191  1.00  0.00           O\\nATOM    421   N GLU A 106      -3.629 -22.313 -27.147  1.00  0.00           N\\nATOM    422 CA  GLU A 106      -3.606 -23.685 -27.646  1.00  0.00           C\\nATOM    423   C GLU A 106      -4.951 -24.075 -28.252  1.00  0.00           C\\nATOM    424   O GLU A 106      -5.001 -24.695 -29.316  1.00  0.00           O\\nATOM    425   N ASN A 107      -6.100 -23.604 -27.645  1.00  0.00           N\\nATOM    426 CA  ASN A 107      -7.441 -23.945 -28.106  1.00  0.00           C\\nATOM    427   C ASN A 107      -7.780 -23.240 -29.417  1.00  0.00           C\\nATOM    428   O ASN A 107      -8.415 -23.826 -30.295  1.00  0.00           O\\nATOM    429   N TYR A 108      -7.183 -21.998 -29.659  1.00  0.00           N\\nATOM    430 CA  TYR A 108      -7.414 -21.255 -30.893  1.00  0.00           C\\nATOM    431   C TYR A 108      -6.576 -21.820 -32.034  1.00  0.00           C\\nATOM    432   O TYR A 108      -6.977 -21.753 -33.198  1.00  0.00           O\\nATOM    433   N CYS A 109      -5.335 -22.397 -31.767  1.00  0.00           N\\nATOM    434 CA  CYS A 109      -4.467 -23.065 -32.730  1.00  0.00           C\\nATOM    435   C CYS A 109      -5.088 -24.373 -33.208  1.00  0.00           C\\nATOM    436   O CYS A 109      -4.905 -24.766 -34.361  1.00  0.00           O\\nATOM    437   N ASN A 110      -5.952 -25.004 -32.405  1.00  0.00           N\\nATOM    438 CA  ASN A 110      -6.677 -26.192 -32.842  1.00  0.00           C\\nATOM    439   C ASN A 110      -7.916 -25.825 -33.655  1.00  0.00           C\\nATOM    440   O ASN A 110      -8.204 -26.457 -34.673  1.00  0.00           O\\nTER\\nEND\\n\",\"pdb\");\n",
       "\tviewer_17613512088053458.setBackgroundColor(\"white\");\n",
       "\tviewer_17613512088053458.setStyle({\"cartoon\": {\"color\": \"spectrum\", \"thickness\": 1.2, \"arrows\": true}});\n",
       "\tviewer_17613512088053458.addStyle({\"atom\": [\"N\", \"CA\", \"C\", \"O\"]},{\"stick\": {\"radius\": 0.25}});\n",
       "\tviewer_17613512088053458.zoomTo();\n",
       "\tviewer_17613512088053458.zoom(1.25);\n",
       "viewer_17613512088053458.render();\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import py3Dmol\n",
    "\n",
    "# --- 1. Define the Correct Vocabulary Mappings ---\n",
    "\n",
    "# This vocabulary list MUST match the one used to create the 'aa' tensor.\n",
    "# It maps the integer index to the one-letter amino acid code.\n",
    "AF2_VOCAB = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', 'X']\n",
    "\n",
    "# This dictionary maps the one-letter code to the three-letter PDB format code.\n",
    "AA1_TO_AA3 = {\n",
    "    'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'D': 'ASP', 'C': 'CYS',\n",
    "    'Q': 'GLN', 'E': 'GLU', 'G': 'GLY', 'H': 'HIS', 'I': 'ILE',\n",
    "    'L': 'LEU', 'K': 'LYS', 'M': 'MET', 'F': 'PHE', 'P': 'PRO',\n",
    "    'S': 'SER', 'T': 'THR', 'W': 'TRP', 'Y': 'TYR', 'V': 'VAL',\n",
    "    'X': 'UNK'  # Map for unknown residues\n",
    "}\n",
    "\n",
    "\n",
    "# --- 2. Extract Data from Model Output ---\n",
    "# (This part assumes 'output' and 'aa' exist from the previous step)\n",
    "\n",
    "LOOK = \"ribbon_plus_sticks\"   # \"bold_ribbon\" | \"ribbon_plus_sticks\" | \"ca_beads\"\n",
    "BIDX = 0 # Batch index to visualize\n",
    "\n",
    "# Get atomic positions and amino acid indices from the tensors\n",
    "pos37 = output[\"positions\"][-1, BIDX].detach().cpu().numpy()\n",
    "aa_idx = aa[BIDX].detach().cpu().numpy()\n",
    "\n",
    "print(\"pos37:\", pos37.shape, \"aa_idx:\", aa_idx.shape, \"NaNs:\", np.isnan(pos37).any())\n",
    "\n",
    "\n",
    "# --- 3. Generate PDB String with Correct Residue Names ---\n",
    "\n",
    "# CORRECTED: Convert indices to three-letter names using the new vocabulary\n",
    "resnames = [AA1_TO_AA3.get(AF2_VOCAB[i], \"UNK\") for i in aa_idx]\n",
    "\n",
    "# Define atom indices for the backbone\n",
    "IDX_N, IDX_CA, IDX_C, IDX_O = 0, 1, 2, 3\n",
    "N, CA, C, O = pos37[:,IDX_N,:], pos37[:,IDX_CA,:], pos37[:,IDX_C,:], pos37[:,IDX_O,:]\n",
    "\n",
    "def pdb_line(serial, name, resname, chain, resseq, xyz, occ=1.00, b=0.00, element=None):\n",
    "    x,y,z = map(float, xyz)\n",
    "    if element is None: element = name[0]\n",
    "    atom_field = f\" {name:>3}\" if len(name)==1 else f\" {name:<3}\"\n",
    "    return (f\"ATOM  {serial:5d}{atom_field}{resname:>4} A{resseq:4d}    \"\n",
    "            f\"{x:8.3f}{y:8.3f}{z:8.3f}{occ:6.2f}{b:6.2f}          {element:>2}\\n\")\n",
    "\n",
    "# Build the PDB file string\n",
    "lines, serial = [], 1\n",
    "for i, resname in enumerate(resnames, start=1):\n",
    "    if i > len(N): break # Stop if we run out of coordinate data (e.g., due to padding)\n",
    "    lines.append(pdb_line(serial, \"N\",  resname, \"A\", i, N[i-1],  element=\"N\")); serial+=1\n",
    "    lines.append(pdb_line(serial, \"CA\", resname, \"A\", i, CA[i-1], element=\"C\")); serial+=1\n",
    "    lines.append(pdb_line(serial, \"C\",  resname, \"A\", i, C[i-1],  element=\"C\")); serial+=1\n",
    "    lines.append(pdb_line(serial, \"O\",  resname, \"A\", i, O[i-1],  element=\"O\")); serial+=1\n",
    "lines.append(\"TER\\nEND\\n\")\n",
    "pdb_str = \"\".join(lines)\n",
    "\n",
    "\n",
    "# --- 4. Visualize with py3Dmol ---\n",
    "\n",
    "v = py3Dmol.view(width=900, height=600)\n",
    "v.addModel(pdb_str, 'pdb')\n",
    "v.setBackgroundColor('white')\n",
    "\n",
    "if LOOK == \"bold_ribbon\":\n",
    "    v.setStyle({'cartoon': {'color':'spectrum', 'thickness':1.6, 'arrows': True}})\n",
    "    v.addStyle({}, {'line': {'linewidth':2}})\n",
    "elif LOOK == \"ribbon_plus_sticks\":\n",
    "    v.setStyle({'cartoon': {'color':'spectrum', 'thickness':1.2, 'arrows': True}})\n",
    "    v.addStyle({'atom':['N','CA','C','O']}, {'stick': {'radius':0.25}})\n",
    "elif LOOK == \"ca_beads\":\n",
    "    v.setStyle({'line': {'linewidth':2}})\n",
    "    v.addStyle({'atom':'CA'}, {'sphere': {'radius':0.6, 'color':'spectrum'}})\n",
    "else:\n",
    "    v.setStyle({'cartoon': {'color':'spectrum'}})\n",
    "\n",
    "v.zoomTo()\n",
    "v.zoom(1.25)\n",
    "v.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Protein-folding-acceleration",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
